<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!-- NewPage -->
<html lang="en">
<head>
<!-- Generated by javadoc -->
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<title>Peptide (computational omics utilities 5.0.1-SNAPSHOT API)</title>
<link rel="stylesheet" type="text/css" href="../../../../../../stylesheet.css" title="Style">
<script type="text/javascript" src="../../../../../../script.js"></script>
</head>
<body>
<script type="text/javascript"><!--
    try {
        if (location.href.indexOf('is-external=true') == -1) {
            parent.document.title="Peptide (computational omics utilities 5.0.1-SNAPSHOT API)";
        }
    }
    catch(err) {
    }
//-->
var methods = {"i0":10,"i1":10,"i2":10,"i3":10,"i4":10,"i5":10,"i6":9,"i7":10,"i8":10,"i9":10,"i10":10,"i11":10,"i12":10,"i13":10,"i14":10,"i15":10,"i16":10,"i17":10,"i18":10,"i19":10,"i20":10,"i21":10,"i22":10,"i23":10,"i24":10,"i25":10,"i26":10,"i27":10,"i28":10,"i29":10,"i30":10,"i31":10,"i32":10,"i33":10,"i34":10,"i35":10,"i36":10,"i37":10,"i38":10,"i39":10};
var tabs = {65535:["t0","All Methods"],1:["t1","Static Methods"],2:["t2","Instance Methods"],8:["t4","Concrete Methods"]};
var altColor = "altColor";
var rowColor = "rowColor";
var tableTab = "tableTab";
var activeTableTab = "activeTableTab";
</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
<!-- ========= START OF TOP NAVBAR ======= -->
<div class="topNav"><a name="navbar.top">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.top" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.top.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../../../../../../overview-summary.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="navBarCell1Rev">Class</li>
<li><a href="class-use/Peptide.html">Use</a></li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../../../../../../deprecated-list.html">Deprecated</a></li>
<li><a href="../../../../../../index-all.html">Index</a></li>
<li><a href="../../../../../../help-doc.html">Help</a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li>Prev&nbsp;Class</li>
<li><a href="../../../../../../com/compomics/util/experiment/biology/proteins/Protein.html" title="class in com.compomics.util.experiment.biology.proteins"><span class="typeNameLink">Next&nbsp;Class</span></a></li>
</ul>
<ul class="navList">
<li><a href="../../../../../../index.html?com/compomics/util/experiment/biology/proteins/Peptide.html" target="_top">Frames</a></li>
<li><a href="Peptide.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_top">
<li><a href="../../../../../../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_top");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<div>
<ul class="subNavList">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li><a href="#field.summary">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.summary">Method</a></li>
</ul>
<ul class="subNavList">
<li>Detail:&nbsp;</li>
<li><a href="#field.detail">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.detail">Method</a></li>
</ul>
</div>
<a name="skip.navbar.top">
<!--   -->
</a></div>
<!-- ========= END OF TOP NAVBAR ========= -->
<!-- ======== START OF CLASS DATA ======== -->
<div class="header">
<div class="subTitle">com.compomics.util.experiment.biology.proteins</div>
<h2 title="Class Peptide" class="title">Class Peptide</h2>
</div>
<div class="contentContainer">
<ul class="inheritance">
<li><a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Object.html?is-external=true" title="class or interface in java.lang">java.lang.Object</a></li>
<li>
<ul class="inheritance">
<li>org.zoodb.api.impl.ZooPC</li>
<li>
<ul class="inheritance">
<li><a href="../../../../../../com/compomics/util/db/object/DbObject.html" title="class in com.compomics.util.db.object">com.compomics.util.db.object.DbObject</a></li>
<li>
<ul class="inheritance">
<li><a href="../../../../../../com/compomics/util/experiment/personalization/ExperimentObject.html" title="class in com.compomics.util.experiment.personalization">com.compomics.util.experiment.personalization.ExperimentObject</a></li>
<li>
<ul class="inheritance">
<li>com.compomics.util.experiment.biology.proteins.Peptide</li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
<div class="description">
<ul class="blockList">
<li class="blockList">
<dl>
<dt>All Implemented Interfaces:</dt>
<dd><a href="http://docs.oracle.com/javase/7/docs/api/java/io/Serializable.html?is-external=true" title="class or interface in java.io">Serializable</a></dd>
</dl>
<hr>
<br>
<pre>public class <span class="typeNameLabel">Peptide</span>
extends <a href="../../../../../../com/compomics/util/experiment/personalization/ExperimentObject.html" title="class in com.compomics.util.experiment.personalization">ExperimentObject</a></pre>
<div class="block">This class represents a peptide. Note that maps and lists provided in
 constructors are used in the class and not cloned.</div>
<dl>
<dt><span class="simpleTagLabel">Author:</span></dt>
<dd>Marc Vaudel, Dominik Kopczynski</dd>
<dt><span class="seeLabel">See Also:</span></dt>
<dd><a href="../../../../../../serialized-form.html#com.compomics.util.experiment.biology.proteins.Peptide">Serialized Form</a></dd>
</dl>
</li>
</ul>
</div>
<div class="summary">
<ul class="blockList">
<li class="blockList">
<!-- =========== FIELD SUMMARY =========== -->
<ul class="blockList">
<li class="blockList"><a name="field.summary">
<!--   -->
</a>
<h3>Field Summary</h3>
<table class="memberSummary" border="0" cellpadding="3" cellspacing="0" summary="Field Summary table, listing fields, and an explanation">
<caption><span>Fields</span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Field and Description</th>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static <a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a></code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html#MODIFICATION_LOCALIZATION_SEPARATOR">MODIFICATION_LOCALIZATION_SEPARATOR</a></span></code>
<div class="block">Separator preceding confident localization of the confident localization
 of a modification.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static <a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a></code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html#MODIFICATION_SEPARATOR">MODIFICATION_SEPARATOR</a></span></code>
<div class="block">Separator used to separate modifications in peptide keys as string.</div>
</td>
</tr>
</table>
</li>
</ul>
<!-- ======== CONSTRUCTOR SUMMARY ======== -->
<ul class="blockList">
<li class="blockList"><a name="constructor.summary">
<!--   -->
</a>
<h3>Constructor Summary</h3>
<table class="memberSummary" border="0" cellpadding="3" cellspacing="0" summary="Constructor Summary table, listing constructors, and an explanation">
<caption><span>Constructors</span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colOne" scope="col">Constructor and Description</th>
</tr>
<tr class="altColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html#Peptide--">Peptide</a></span>()</code>
<div class="block">Constructor for the peptide.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html#Peptide-java.lang.String-">Peptide</a></span>(<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;aSequence)</code>
<div class="block">Constructor.</div>
</td>
</tr>
<tr class="altColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html#Peptide-java.lang.String-com.compomics.util.experiment.identification.matches.ModificationMatch:A-">Peptide</a></span>(<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;aSequence,
       <a href="../../../../../../com/compomics/util/experiment/identification/matches/ModificationMatch.html" title="class in com.compomics.util.experiment.identification.matches">ModificationMatch</a>[]&nbsp;modifications)</code>
<div class="block">Constructor.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html#Peptide-java.lang.String-com.compomics.util.experiment.identification.matches.ModificationMatch:A-boolean-">Peptide</a></span>(<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;aSequence,
       <a href="../../../../../../com/compomics/util/experiment/identification/matches/ModificationMatch.html" title="class in com.compomics.util.experiment.identification.matches">ModificationMatch</a>[]&nbsp;modificationMatches,
       boolean&nbsp;sanityCheck)</code>
<div class="block">Constructor.</div>
</td>
</tr>
<tr class="altColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html#Peptide-java.lang.String-com.compomics.util.experiment.identification.matches.ModificationMatch:A-boolean-double-">Peptide</a></span>(<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;aSequence,
       <a href="../../../../../../com/compomics/util/experiment/identification/matches/ModificationMatch.html" title="class in com.compomics.util.experiment.identification.matches">ModificationMatch</a>[]&nbsp;modificationMatches,
       boolean&nbsp;sanityCheck,
       double&nbsp;mass)</code>
<div class="block">Constructor.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html#Peptide-java.lang.String-com.compomics.util.experiment.identification.matches.ModificationMatch:A-java.util.HashMap-boolean-">Peptide</a></span>(<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;aSequence,
       <a href="../../../../../../com/compomics/util/experiment/identification/matches/ModificationMatch.html" title="class in com.compomics.util.experiment.identification.matches">ModificationMatch</a>[]&nbsp;modificationMatches,
       <a href="http://docs.oracle.com/javase/7/docs/api/java/util/HashMap.html?is-external=true" title="class or interface in java.util">HashMap</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>,<a href="http://docs.oracle.com/javase/7/docs/api/java/util/HashMap.html?is-external=true" title="class or interface in java.util">HashMap</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Integer.html?is-external=true" title="class or interface in java.lang">Integer</a>,<a href="../../../../../../com/compomics/util/experiment/identification/matches/PeptideVariantMatches.html" title="class in com.compomics.util.experiment.identification.matches">PeptideVariantMatches</a>&gt;&gt;&nbsp;variantMatches,
       boolean&nbsp;sanityCheck)</code>
<div class="block">Constructor for the peptide.</div>
</td>
</tr>
<tr class="altColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html#Peptide-java.lang.String-com.compomics.util.experiment.identification.matches.ModificationMatch:A-java.util.HashMap-boolean-double-">Peptide</a></span>(<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;aSequence,
       <a href="../../../../../../com/compomics/util/experiment/identification/matches/ModificationMatch.html" title="class in com.compomics.util.experiment.identification.matches">ModificationMatch</a>[]&nbsp;modificationMatches,
       <a href="http://docs.oracle.com/javase/7/docs/api/java/util/HashMap.html?is-external=true" title="class or interface in java.util">HashMap</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>,<a href="http://docs.oracle.com/javase/7/docs/api/java/util/HashMap.html?is-external=true" title="class or interface in java.util">HashMap</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Integer.html?is-external=true" title="class or interface in java.lang">Integer</a>,<a href="../../../../../../com/compomics/util/experiment/identification/matches/PeptideVariantMatches.html" title="class in com.compomics.util.experiment.identification.matches">PeptideVariantMatches</a>&gt;&gt;&nbsp;variantMatches,
       boolean&nbsp;sanityCheck,
       double&nbsp;mass)</code>
<div class="block">Constructor.</div>
</td>
</tr>
</table>
</li>
</ul>
<!-- ========== METHOD SUMMARY =========== -->
<ul class="blockList">
<li class="blockList"><a name="method.summary">
<!--   -->
</a>
<h3>Method Summary</h3>
<table class="memberSummary" border="0" cellpadding="3" cellspacing="0" summary="Method Summary table, listing methods, and an explanation">
<caption><span id="t0" class="activeTableTab"><span>All Methods</span><span class="tabEnd">&nbsp;</span></span><span id="t1" class="tableTab"><span><a href="javascript:show(1);">Static Methods</a></span><span class="tabEnd">&nbsp;</span></span><span id="t2" class="tableTab"><span><a href="javascript:show(2);">Instance Methods</a></span><span class="tabEnd">&nbsp;</span></span><span id="t4" class="tableTab"><span><a href="javascript:show(8);">Concrete Methods</a></span><span class="tabEnd">&nbsp;</span></span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tr id="i0" class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html#addModificationMatch-com.compomics.util.experiment.identification.matches.ModificationMatch-">addModificationMatch</a></span>(<a href="../../../../../../com/compomics/util/experiment/identification/matches/ModificationMatch.html" title="class in com.compomics.util.experiment.identification.matches">ModificationMatch</a>&nbsp;modificationMatch)</code>
<div class="block">Adds a modification match.</div>
</td>
</tr>
<tr id="i1" class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html#clearModificationMatches--">clearModificationMatches</a></span>()</code>
<div class="block">Clears the list of imported modification matches.</div>
</td>
</tr>
<tr id="i2" class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html#clearVariantMatches--">clearVariantMatches</a></span>()</code>
<div class="block">Clears the list of imported variant matches.</div>
</td>
</tr>
<tr id="i3" class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html#estimateTheoreticMass--">estimateTheoreticMass</a></span>()</code>
<div class="block">Estimates the theoretic mass of the peptide.</div>
</td>
</tr>
<tr id="i4" class="altColor">
<td class="colFirst"><code><a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a></code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html#getCTerminal--">getCTerminal</a></span>()</code>
<div class="block">Returns the C-terminal of the peptide as a String.</div>
</td>
</tr>
<tr id="i5" class="rowColor">
<td class="colFirst"><code>long</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html#getKey--">getKey</a></span>()</code>
<div class="block">Returns the reference key of a peptide.</div>
</td>
</tr>
<tr id="i6" class="altColor">
<td class="colFirst"><code>static long</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html#getKey-java.lang.String-com.compomics.util.experiment.identification.matches.ModificationMatch:A-">getKey</a></span>(<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;sequence,
      <a href="../../../../../../com/compomics/util/experiment/identification/matches/ModificationMatch.html" title="class in com.compomics.util.experiment.identification.matches">ModificationMatch</a>[]&nbsp;modificationMatches)</code>
<div class="block">Returns the reference key of a peptide.</div>
</td>
</tr>
<tr id="i7" class="rowColor">
<td class="colFirst"><code>double</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html#getMass--">getMass</a></span>()</code>
<div class="block">Getter for the mass.</div>
</td>
</tr>
<tr id="i8" class="altColor">
<td class="colFirst"><code>long</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html#getMatchingKey--">getMatchingKey</a></span>()</code>
<div class="block">Returns the key accounting for sequence matching preferences</div>
</td>
</tr>
<tr id="i9" class="rowColor">
<td class="colFirst"><code>long</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html#getMatchingKey-com.compomics.util.parameters.identification.advanced.SequenceMatchingParameters-">getMatchingKey</a></span>(<a href="../../../../../../com/compomics/util/parameters/identification/advanced/SequenceMatchingParameters.html" title="class in com.compomics.util.parameters.identification.advanced">SequenceMatchingParameters</a>&nbsp;sequenceMatchingPreferences)</code>
<div class="block">Returns a unique key for the peptide when considering the given matching
 preferences.</div>
</td>
</tr>
<tr id="i10" class="altColor">
<td class="colFirst"><code><a href="../../../../../../com/compomics/util/experiment/identification/matches/ModificationMatch.html" title="class in com.compomics.util.experiment.identification.matches">ModificationMatch</a>[]</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html#getModificationMatches--">getModificationMatches</a></span>()</code>
<div class="block">Getter for the modifications carried by this peptide.</div>
</td>
</tr>
<tr id="i11" class="rowColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html#getNMissedCleavages-com.compomics.util.parameters.identification.search.DigestionParameters-">getNMissedCleavages</a></span>(<a href="../../../../../../com/compomics/util/parameters/identification/search/DigestionParameters.html" title="class in com.compomics.util.parameters.identification.search">DigestionParameters</a>&nbsp;digestionPreferences)</code>
<div class="block">Returns the number of missed cleavages using the digestion preferences.</div>
</td>
</tr>
<tr id="i12" class="altColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html#getNMissedCleavages-com.compomics.util.experiment.biology.enzymes.Enzyme-">getNMissedCleavages</a></span>(<a href="../../../../../../com/compomics/util/experiment/biology/enzymes/Enzyme.html" title="class in com.compomics.util.experiment.biology.enzymes">Enzyme</a>&nbsp;enzyme)</code>
<div class="block">Returns the number of missed cleavages using the specified enzyme.</div>
</td>
</tr>
<tr id="i13" class="rowColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html#getNModifications--">getNModifications</a></span>()</code>
<div class="block">Returns the number of modifications carried by this peptide.</div>
</td>
</tr>
<tr id="i14" class="altColor">
<td class="colFirst"><code><a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a></code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html#getNoModPeptide-java.util.HashSet-">getNoModPeptide</a></span>(<a href="http://docs.oracle.com/javase/7/docs/api/java/util/HashSet.html?is-external=true" title="class or interface in java.util">HashSet</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&gt;&nbsp;forbiddenModifications)</code>
<div class="block">Returns a version of the peptide which does not contain the given list of
 modifications.</div>
</td>
</tr>
<tr id="i15" class="rowColor">
<td class="colFirst"><code><a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a></code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html#getNTerminal--">getNTerminal</a></span>()</code>
<div class="block">Returns the N-terminal of the peptide as a String.</div>
</td>
</tr>
<tr id="i16" class="altColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html#getNVariableModifications-double-">getNVariableModifications</a></span>(double&nbsp;modificationMass)</code>
<div class="block">Returns the number of variable modifications found with the given mass.</div>
</td>
</tr>
<tr id="i17" class="rowColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html#getPeptideEnd-java.lang.String-int-">getPeptideEnd</a></span>(<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;proteinAccession,
             int&nbsp;peptideStart)</code>
<div class="block">Returns the 0 based end index of the peptide on the protein sequence.</div>
</td>
</tr>
<tr id="i18" class="altColor">
<td class="colFirst"><code><a href="http://docs.oracle.com/javase/7/docs/api/java/util/HashSet.html?is-external=true" title="class or interface in java.util">HashSet</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Integer.html?is-external=true" title="class or interface in java.lang">Integer</a>&gt;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html#getPotentialModificationSites-com.compomics.util.experiment.biology.modifications.Modification-com.compomics.util.experiment.io.biology.protein.SequenceProvider-com.compomics.util.parameters.identification.advanced.SequenceMatchingParameters-">getPotentialModificationSites</a></span>(<a href="../../../../../../com/compomics/util/experiment/biology/modifications/Modification.html" title="class in com.compomics.util.experiment.biology.modifications">Modification</a>&nbsp;modification,
                             <a href="../../../../../../com/compomics/util/experiment/io/biology/protein/SequenceProvider.html" title="interface in com.compomics.util.experiment.io.biology.protein">SequenceProvider</a>&nbsp;sequenceProvider,
                             <a href="../../../../../../com/compomics/util/parameters/identification/advanced/SequenceMatchingParameters.html" title="class in com.compomics.util.parameters.identification.advanced">SequenceMatchingParameters</a>&nbsp;modificationSequenceMatchingPreferences)</code>
<div class="block">Returns the potential modification sites as a set.</div>
</td>
</tr>
<tr id="i19" class="rowColor">
<td class="colFirst"><code><a href="http://docs.oracle.com/javase/7/docs/api/java/util/ArrayList.html?is-external=true" title="class or interface in java.util">ArrayList</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Integer.html?is-external=true" title="class or interface in java.lang">Integer</a>&gt;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html#getPotentialModificationSites-com.compomics.util.experiment.biology.modifications.Modification-java.lang.String-int-com.compomics.util.parameters.identification.advanced.SequenceMatchingParameters-">getPotentialModificationSites</a></span>(<a href="../../../../../../com/compomics/util/experiment/biology/modifications/Modification.html" title="class in com.compomics.util.experiment.biology.modifications">Modification</a>&nbsp;modification,
                             <a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;proteinSequence,
                             int&nbsp;peptideStart,
                             <a href="../../../../../../com/compomics/util/parameters/identification/advanced/SequenceMatchingParameters.html" title="class in com.compomics.util.parameters.identification.advanced">SequenceMatchingParameters</a>&nbsp;modificationSequenceMatchingPreferences)</code>
<div class="block">Returns the potential modification sites as an ordered list of sites.</div>
</td>
</tr>
<tr id="i20" class="altColor">
<td class="colFirst"><code><a href="http://docs.oracle.com/javase/7/docs/api/java/util/ArrayList.html?is-external=true" title="class or interface in java.util">ArrayList</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Integer.html?is-external=true" title="class or interface in java.lang">Integer</a>&gt;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html#getPotentialModificationSitesNoCombination-com.compomics.util.experiment.biology.modifications.Modification-java.lang.String-int-">getPotentialModificationSitesNoCombination</a></span>(<a href="../../../../../../com/compomics/util/experiment/biology/modifications/Modification.html" title="class in com.compomics.util.experiment.biology.modifications">Modification</a>&nbsp;modification,
                                          <a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;proteinSequence,
                                          int&nbsp;peptideStart)</code>
<div class="block">Returns the potential modification sites as an ordered list of sites.</div>
</td>
</tr>
<tr id="i21" class="rowColor">
<td class="colFirst"><code><a href="http://docs.oracle.com/javase/7/docs/api/java/util/TreeMap.html?is-external=true" title="class or interface in java.util">TreeMap</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>,int[]&gt;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html#getProteinMapping--">getProteinMapping</a></span>()</code>
<div class="block">Returns the proteins mapping as a map of 0 based indexes for every
 protein accession.</div>
</td>
</tr>
<tr id="i22" class="altColor">
<td class="colFirst"><code><a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a></code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html#getSequence--">getSequence</a></span>()</code>
<div class="block">Returns for the sequence.</div>
</td>
</tr>
<tr id="i23" class="rowColor">
<td class="colFirst"><code><a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a></code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html#getTaggedModifiedSequence-com.compomics.util.parameters.identification.search.ModificationParameters-boolean-boolean-boolean-">getTaggedModifiedSequence</a></span>(<a href="../../../../../../com/compomics/util/parameters/identification/search/ModificationParameters.html" title="class in com.compomics.util.parameters.identification.search">ModificationParameters</a>&nbsp;modificationProfile,
                         boolean&nbsp;useHtmlColorCoding,
                         boolean&nbsp;includeHtmlStartEndTags,
                         boolean&nbsp;useShortName)</code>
<div class="block">Returns the modified sequence as an tagged string with potential
 modification sites color coded or with Modification tags, e.g,
 &lt;mox&gt;.</div>
</td>
</tr>
<tr id="i24" class="altColor">
<td class="colFirst"><code><a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a></code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html#getTaggedModifiedSequence-com.compomics.util.parameters.identification.search.ModificationParameters-boolean-boolean-boolean-boolean-">getTaggedModifiedSequence</a></span>(<a href="../../../../../../com/compomics/util/parameters/identification/search/ModificationParameters.html" title="class in com.compomics.util.parameters.identification.search">ModificationParameters</a>&nbsp;modificationProfile,
                         boolean&nbsp;useHtmlColorCoding,
                         boolean&nbsp;includeHtmlStartEndTags,
                         boolean&nbsp;useShortName,
                         boolean&nbsp;excludeAllFixedPtms)</code>
<div class="block">Returns the modified sequence as an tagged string with potential
 modification sites color coded or with Modification tags, e.g,
 &lt;mox&gt;.</div>
</td>
</tr>
<tr id="i25" class="rowColor">
<td class="colFirst"><code><a href="http://docs.oracle.com/javase/7/docs/api/java/util/HashMap.html?is-external=true" title="class or interface in java.util">HashMap</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>,<a href="http://docs.oracle.com/javase/7/docs/api/java/util/HashMap.html?is-external=true" title="class or interface in java.util">HashMap</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Integer.html?is-external=true" title="class or interface in java.lang">Integer</a>,<a href="../../../../../../com/compomics/util/experiment/identification/matches/PeptideVariantMatches.html" title="class in com.compomics.util.experiment.identification.matches">PeptideVariantMatches</a>&gt;&gt;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html#getVariantMatches--">getVariantMatches</a></span>()</code>
<div class="block">Returns the sequence variant matches of this peptide indexed by protein
 accession and peptide start.</div>
</td>
</tr>
<tr id="i26" class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html#isCterm-java.lang.String-int-">isCterm</a></span>(<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;proteinSequence,
       int&nbsp;peptideStart)</code>
<div class="block">Returns a boolean indicating whether the peptide is at the C-terminus of
 the given protein sequence.</div>
</td>
</tr>
<tr id="i27" class="rowColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html#isNterm-java.lang.String-int-">isNterm</a></span>(<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;proteinSequence,
       int&nbsp;peptideStart)</code>
<div class="block">Returns a boolean indicating whether the peptide is at the N-terminus of
 the given protein sequence.</div>
</td>
</tr>
<tr id="i28" class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html#isSameModificationStatus-com.compomics.util.experiment.biology.proteins.Peptide-">isSameModificationStatus</a></span>(<a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;anotherPeptide)</code>
<div class="block">Indicates whether another peptide has the same variable modifications as
 this peptide.</div>
</td>
</tr>
<tr id="i29" class="rowColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html#isSameSequence-com.compomics.util.experiment.biology.proteins.Peptide-com.compomics.util.parameters.identification.advanced.SequenceMatchingParameters-">isSameSequence</a></span>(<a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;anotherPeptide,
              <a href="../../../../../../com/compomics/util/parameters/identification/advanced/SequenceMatchingParameters.html" title="class in com.compomics.util.parameters.identification.advanced">SequenceMatchingParameters</a>&nbsp;sequenceMatchingPreferences)</code>
<div class="block">Returns a boolean indicating whether another peptide has the same
 sequence as the given peptide</div>
</td>
</tr>
<tr id="i30" class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html#isSameSequenceAndModificationStatus-com.compomics.util.experiment.biology.proteins.Peptide-com.compomics.util.parameters.identification.advanced.SequenceMatchingParameters-">isSameSequenceAndModificationStatus</a></span>(<a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;anotherPeptide,
                                   <a href="../../../../../../com/compomics/util/parameters/identification/advanced/SequenceMatchingParameters.html" title="class in com.compomics.util.parameters.identification.advanced">SequenceMatchingParameters</a>&nbsp;sequenceMatchingPreferences)</code>
<div class="block">Indicates whether another peptide has the same sequence and modification
 status without accounting for modification localization.</div>
</td>
</tr>
<tr id="i31" class="rowColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html#sameModificationsAs-com.compomics.util.experiment.biology.proteins.Peptide-">sameModificationsAs</a></span>(<a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;anotherPeptide)</code>
<div class="block">Indicates whether another peptide has the same modifications at the same
 localization as this peptide.</div>
</td>
</tr>
<tr id="i32" class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html#sameModificationsAs-com.compomics.util.experiment.biology.proteins.Peptide-java.util.ArrayList-">sameModificationsAs</a></span>(<a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;anotherPeptide,
                   <a href="http://docs.oracle.com/javase/7/docs/api/java/util/ArrayList.html?is-external=true" title="class or interface in java.util">ArrayList</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&gt;&nbsp;modifications)</code>
<div class="block">Indicates whether another peptide has the same modifications at the same
 localization as this peptide.</div>
</td>
</tr>
<tr id="i33" class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html#setKey-long-">setKey</a></span>(long&nbsp;key)</code>
<div class="block">Sets the object key.</div>
</td>
</tr>
<tr id="i34" class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html#setMass-double-">setMass</a></span>(double&nbsp;mass)</code>
<div class="block">Sets the mass.</div>
</td>
</tr>
<tr id="i35" class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html#setMatchingKey-long-">setMatchingKey</a></span>(long&nbsp;matchingKey)</code>
<div class="block">Sets the key accounting for sequence matching preferences.</div>
</td>
</tr>
<tr id="i36" class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html#setModificationMatches-com.compomics.util.experiment.identification.matches.ModificationMatch:A-">setModificationMatches</a></span>(<a href="../../../../../../com/compomics/util/experiment/identification/matches/ModificationMatch.html" title="class in com.compomics.util.experiment.identification.matches">ModificationMatch</a>[]&nbsp;modificationMatches)</code>
<div class="block">Sets new modification matches for the peptide.</div>
</td>
</tr>
<tr id="i37" class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html#setProteinMapping-java.util.TreeMap-">setProteinMapping</a></span>(<a href="http://docs.oracle.com/javase/7/docs/api/java/util/TreeMap.html?is-external=true" title="class or interface in java.util">TreeMap</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>,int[]&gt;&nbsp;proteinMapping)</code>
<div class="block">Sets the proteins mapping as a map of 0 based indexes for every protein
 accession.</div>
</td>
</tr>
<tr id="i38" class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html#setSequence-java.lang.String-">setSequence</a></span>(<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;sequence)</code>
<div class="block">Sets for the sequence.</div>
</td>
</tr>
<tr id="i39" class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html#setVariantMatches-java.util.HashMap-">setVariantMatches</a></span>(<a href="http://docs.oracle.com/javase/7/docs/api/java/util/HashMap.html?is-external=true" title="class or interface in java.util">HashMap</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>,<a href="http://docs.oracle.com/javase/7/docs/api/java/util/HashMap.html?is-external=true" title="class or interface in java.util">HashMap</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Integer.html?is-external=true" title="class or interface in java.lang">Integer</a>,<a href="../../../../../../com/compomics/util/experiment/identification/matches/PeptideVariantMatches.html" title="class in com.compomics.util.experiment.identification.matches">PeptideVariantMatches</a>&gt;&gt;&nbsp;variantMatches)</code>
<div class="block">Sets the sequence variant matches of this peptide.</div>
</td>
</tr>
</table>
<ul class="blockList">
<li class="blockList"><a name="methods.inherited.from.class.com.compomics.util.experiment.personalization.ExperimentObject">
<!--   -->
</a>
<h3>Methods inherited from class&nbsp;com.compomics.util.experiment.personalization.<a href="../../../../../../com/compomics/util/experiment/personalization/ExperimentObject.html" title="class in com.compomics.util.experiment.personalization">ExperimentObject</a></h3>
<code><a href="../../../../../../com/compomics/util/experiment/personalization/ExperimentObject.html#addUrParam-com.compomics.util.experiment.personalization.UrParameter-">addUrParam</a>, <a href="../../../../../../com/compomics/util/experiment/personalization/ExperimentObject.html#asLong-java.lang.String-">asLong</a>, <a href="../../../../../../com/compomics/util/experiment/personalization/ExperimentObject.html#clearParametersMap--">clearParametersMap</a>, <a href="../../../../../../com/compomics/util/experiment/personalization/ExperimentObject.html#getUrParam-com.compomics.util.experiment.personalization.UrParameter-">getUrParam</a>, <a href="../../../../../../com/compomics/util/experiment/personalization/ExperimentObject.html#getUrParams--">getUrParams</a>, <a href="../../../../../../com/compomics/util/experiment/personalization/ExperimentObject.html#removeUrParam-long-">removeUrParam</a>, <a href="../../../../../../com/compomics/util/experiment/personalization/ExperimentObject.html#setUrParams-java.util.HashMap-">setUrParams</a></code></li>
</ul>
<ul class="blockList">
<li class="blockList"><a name="methods.inherited.from.class.com.compomics.util.db.object.DbObject">
<!--   -->
</a>
<h3>Methods inherited from class&nbsp;com.compomics.util.db.object.<a href="../../../../../../com/compomics/util/db/object/DbObject.html" title="class in com.compomics.util.db.object">DbObject</a></h3>
<code><a href="../../../../../../com/compomics/util/db/object/DbObject.html#getFirstLevel--">getFirstLevel</a>, <a href="../../../../../../com/compomics/util/db/object/DbObject.html#getId--">getId</a>, <a href="../../../../../../com/compomics/util/db/object/DbObject.html#setFirstLevel-boolean-">setFirstLevel</a>, <a href="../../../../../../com/compomics/util/db/object/DbObject.html#setId-long-">setId</a></code></li>
</ul>
<ul class="blockList">
<li class="blockList"><a name="methods.inherited.from.class.org.zoodb.api.impl.ZooPC">
<!--   -->
</a>
<h3>Methods inherited from class&nbsp;org.zoodb.api.impl.ZooPC</h3>
<code>jdoZooEvict, jdoZooGetBackup, jdoZooGetClassDef, jdoZooGetContext, jdoZooGetNode, jdoZooGetOid, jdoZooGetTimestamp, jdoZooHasState, jdoZooInit, jdoZooIsDeleted, jdoZooIsDetached, jdoZooIsDirty, jdoZooIsNew, jdoZooIsPersistent, jdoZooIsStateHollow, jdoZooIsTransactional, jdoZooMarkClean, jdoZooMarkDeleted, jdoZooMarkDetached, jdoZooMarkDirty, jdoZooMarkHollow, jdoZooMarkTransient, jdoZooSetOid, jdoZooSetTimestamp, toString, zooActivateRead, zooActivateWrite, zooActivateWrite</code></li>
</ul>
<ul class="blockList">
<li class="blockList"><a name="methods.inherited.from.class.java.lang.Object">
<!--   -->
</a>
<h3>Methods inherited from class&nbsp;java.lang.<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Object.html?is-external=true" title="class or interface in java.lang">Object</a></h3>
<code><a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Object.html?is-external=true#clone--" title="class or interface in java.lang">clone</a>, <a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Object.html?is-external=true#equals-java.lang.Object-" title="class or interface in java.lang">equals</a>, <a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Object.html?is-external=true#finalize--" title="class or interface in java.lang">finalize</a>, <a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Object.html?is-external=true#getClass--" title="class or interface in java.lang">getClass</a>, <a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Object.html?is-external=true#hashCode--" title="class or interface in java.lang">hashCode</a>, <a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Object.html?is-external=true#notify--" title="class or interface in java.lang">notify</a>, <a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Object.html?is-external=true#notifyAll--" title="class or interface in java.lang">notifyAll</a>, <a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Object.html?is-external=true#wait--" title="class or interface in java.lang">wait</a>, <a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Object.html?is-external=true#wait-long-" title="class or interface in java.lang">wait</a>, <a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Object.html?is-external=true#wait-long-int-" title="class or interface in java.lang">wait</a></code></li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="details">
<ul class="blockList">
<li class="blockList">
<!-- ============ FIELD DETAIL =========== -->
<ul class="blockList">
<li class="blockList"><a name="field.detail">
<!--   -->
</a>
<h3>Field Detail</h3>
<a name="MODIFICATION_LOCALIZATION_SEPARATOR">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>MODIFICATION_LOCALIZATION_SEPARATOR</h4>
<pre>public static final&nbsp;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a> MODIFICATION_LOCALIZATION_SEPARATOR</pre>
<div class="block">Separator preceding confident localization of the confident localization
 of a modification.</div>
<dl>
<dt><span class="seeLabel">See Also:</span></dt>
<dd><a href="../../../../../../constant-values.html#com.compomics.util.experiment.biology.proteins.Peptide.MODIFICATION_LOCALIZATION_SEPARATOR">Constant Field Values</a></dd>
</dl>
</li>
</ul>
<a name="MODIFICATION_SEPARATOR">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>MODIFICATION_SEPARATOR</h4>
<pre>public static final&nbsp;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a> MODIFICATION_SEPARATOR</pre>
<div class="block">Separator used to separate modifications in peptide keys as string.</div>
<dl>
<dt><span class="seeLabel">See Also:</span></dt>
<dd><a href="../../../../../../constant-values.html#com.compomics.util.experiment.biology.proteins.Peptide.MODIFICATION_SEPARATOR">Constant Field Values</a></dd>
</dl>
</li>
</ul>
</li>
</ul>
<!-- ========= CONSTRUCTOR DETAIL ======== -->
<ul class="blockList">
<li class="blockList"><a name="constructor.detail">
<!--   -->
</a>
<h3>Constructor Detail</h3>
<a name="Peptide--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>Peptide</h4>
<pre>public&nbsp;Peptide()</pre>
<div class="block">Constructor for the peptide.</div>
</li>
</ul>
<a name="Peptide-java.lang.String-com.compomics.util.experiment.identification.matches.ModificationMatch:A-java.util.HashMap-boolean-double-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>Peptide</h4>
<pre>public&nbsp;Peptide(<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;aSequence,
               <a href="../../../../../../com/compomics/util/experiment/identification/matches/ModificationMatch.html" title="class in com.compomics.util.experiment.identification.matches">ModificationMatch</a>[]&nbsp;modificationMatches,
               <a href="http://docs.oracle.com/javase/7/docs/api/java/util/HashMap.html?is-external=true" title="class or interface in java.util">HashMap</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>,<a href="http://docs.oracle.com/javase/7/docs/api/java/util/HashMap.html?is-external=true" title="class or interface in java.util">HashMap</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Integer.html?is-external=true" title="class or interface in java.lang">Integer</a>,<a href="../../../../../../com/compomics/util/experiment/identification/matches/PeptideVariantMatches.html" title="class in com.compomics.util.experiment.identification.matches">PeptideVariantMatches</a>&gt;&gt;&nbsp;variantMatches,
               boolean&nbsp;sanityCheck,
               double&nbsp;mass)</pre>
<div class="block">Constructor.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>aSequence</code> - the peptide sequence, assumed to be in upper case only</dd>
<dd><code>modificationMatches</code> - the Modification of this peptide</dd>
<dd><code>variantMatches</code> - the sequence variants compared to the database</dd>
<dd><code>sanityCheck</code> - boolean indicating whether the input should be checked</dd>
<dd><code>mass</code> - the mass of the peptide</dd>
</dl>
</li>
</ul>
<a name="Peptide-java.lang.String-com.compomics.util.experiment.identification.matches.ModificationMatch:A-boolean-double-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>Peptide</h4>
<pre>public&nbsp;Peptide(<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;aSequence,
               <a href="../../../../../../com/compomics/util/experiment/identification/matches/ModificationMatch.html" title="class in com.compomics.util.experiment.identification.matches">ModificationMatch</a>[]&nbsp;modificationMatches,
               boolean&nbsp;sanityCheck,
               double&nbsp;mass)</pre>
<div class="block">Constructor.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>aSequence</code> - the peptide sequence, assumed to be in upper case only</dd>
<dd><code>modificationMatches</code> - the Modification of this peptide</dd>
<dd><code>sanityCheck</code> - boolean indicating whether the input should be checked</dd>
<dd><code>mass</code> - the mass of the peptide</dd>
</dl>
</li>
</ul>
<a name="Peptide-java.lang.String-com.compomics.util.experiment.identification.matches.ModificationMatch:A-boolean-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>Peptide</h4>
<pre>public&nbsp;Peptide(<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;aSequence,
               <a href="../../../../../../com/compomics/util/experiment/identification/matches/ModificationMatch.html" title="class in com.compomics.util.experiment.identification.matches">ModificationMatch</a>[]&nbsp;modificationMatches,
               boolean&nbsp;sanityCheck)</pre>
<div class="block">Constructor.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>aSequence</code> - the peptide sequence, assumed to be in upper case only</dd>
<dd><code>modificationMatches</code> - the Modification of this peptide</dd>
<dd><code>sanityCheck</code> - boolean indicating whether the input should be checked</dd>
</dl>
</li>
</ul>
<a name="Peptide-java.lang.String-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>Peptide</h4>
<pre>public&nbsp;Peptide(<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;aSequence)</pre>
<div class="block">Constructor. No sanity check is performed on the input.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>aSequence</code> - the peptide sequence</dd>
</dl>
</li>
</ul>
<a name="Peptide-java.lang.String-com.compomics.util.experiment.identification.matches.ModificationMatch:A-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>Peptide</h4>
<pre>public&nbsp;Peptide(<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;aSequence,
               <a href="../../../../../../com/compomics/util/experiment/identification/matches/ModificationMatch.html" title="class in com.compomics.util.experiment.identification.matches">ModificationMatch</a>[]&nbsp;modifications)</pre>
<div class="block">Constructor. No sanity check is performed on the input.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>aSequence</code> - the peptide sequence, assumed to be in upper case only</dd>
<dd><code>modifications</code> - the Modification of this peptide</dd>
</dl>
</li>
</ul>
<a name="Peptide-java.lang.String-com.compomics.util.experiment.identification.matches.ModificationMatch:A-java.util.HashMap-boolean-">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>Peptide</h4>
<pre>public&nbsp;Peptide(<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;aSequence,
               <a href="../../../../../../com/compomics/util/experiment/identification/matches/ModificationMatch.html" title="class in com.compomics.util.experiment.identification.matches">ModificationMatch</a>[]&nbsp;modificationMatches,
               <a href="http://docs.oracle.com/javase/7/docs/api/java/util/HashMap.html?is-external=true" title="class or interface in java.util">HashMap</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>,<a href="http://docs.oracle.com/javase/7/docs/api/java/util/HashMap.html?is-external=true" title="class or interface in java.util">HashMap</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Integer.html?is-external=true" title="class or interface in java.lang">Integer</a>,<a href="../../../../../../com/compomics/util/experiment/identification/matches/PeptideVariantMatches.html" title="class in com.compomics.util.experiment.identification.matches">PeptideVariantMatches</a>&gt;&gt;&nbsp;variantMatches,
               boolean&nbsp;sanityCheck)</pre>
<div class="block">Constructor for the peptide.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>aSequence</code> - the peptide sequence, assumed to be in upper case only</dd>
<dd><code>modificationMatches</code> - the modifications of this peptide</dd>
<dd><code>variantMatches</code> - the variants compared to the database</dd>
<dd><code>sanityCheck</code> - boolean indicating whether the input should be checked</dd>
</dl>
</li>
</ul>
</li>
</ul>
<!-- ============ METHOD DETAIL ========== -->
<ul class="blockList">
<li class="blockList"><a name="method.detail">
<!--   -->
</a>
<h3>Method Detail</h3>
<a name="setMass-double-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>setMass</h4>
<pre>public&nbsp;void&nbsp;setMass(double&nbsp;mass)</pre>
<div class="block">Sets the mass.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>mass</code> - the mass</dd>
</dl>
</li>
</ul>
<a name="setKey-long-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>setKey</h4>
<pre>public&nbsp;void&nbsp;setKey(long&nbsp;key)</pre>
<div class="block">Sets the object key.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>key</code> - the object key</dd>
</dl>
</li>
</ul>
<a name="getProteinMapping--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getProteinMapping</h4>
<pre>public&nbsp;<a href="http://docs.oracle.com/javase/7/docs/api/java/util/TreeMap.html?is-external=true" title="class or interface in java.util">TreeMap</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>,int[]&gt;&nbsp;getProteinMapping()</pre>
<div class="block">Returns the proteins mapping as a map of 0 based indexes for every
 protein accession.</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the proteins mapping</dd>
</dl>
</li>
</ul>
<a name="setProteinMapping-java.util.TreeMap-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>setProteinMapping</h4>
<pre>public&nbsp;void&nbsp;setProteinMapping(<a href="http://docs.oracle.com/javase/7/docs/api/java/util/TreeMap.html?is-external=true" title="class or interface in java.util">TreeMap</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>,int[]&gt;&nbsp;proteinMapping)</pre>
<div class="block">Sets the proteins mapping as a map of 0 based indexes for every protein
 accession.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>proteinMapping</code> - the proteins mapping</dd>
</dl>
</li>
</ul>
<a name="setVariantMatches-java.util.HashMap-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>setVariantMatches</h4>
<pre>public&nbsp;void&nbsp;setVariantMatches(<a href="http://docs.oracle.com/javase/7/docs/api/java/util/HashMap.html?is-external=true" title="class or interface in java.util">HashMap</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>,<a href="http://docs.oracle.com/javase/7/docs/api/java/util/HashMap.html?is-external=true" title="class or interface in java.util">HashMap</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Integer.html?is-external=true" title="class or interface in java.lang">Integer</a>,<a href="../../../../../../com/compomics/util/experiment/identification/matches/PeptideVariantMatches.html" title="class in com.compomics.util.experiment.identification.matches">PeptideVariantMatches</a>&gt;&gt;&nbsp;variantMatches)</pre>
<div class="block">Sets the sequence variant matches of this peptide.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>variantMatches</code> - the variant matches of this peptide</dd>
</dl>
</li>
</ul>
<a name="getVariantMatches--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getVariantMatches</h4>
<pre>public&nbsp;<a href="http://docs.oracle.com/javase/7/docs/api/java/util/HashMap.html?is-external=true" title="class or interface in java.util">HashMap</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>,<a href="http://docs.oracle.com/javase/7/docs/api/java/util/HashMap.html?is-external=true" title="class or interface in java.util">HashMap</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Integer.html?is-external=true" title="class or interface in java.lang">Integer</a>,<a href="../../../../../../com/compomics/util/experiment/identification/matches/PeptideVariantMatches.html" title="class in com.compomics.util.experiment.identification.matches">PeptideVariantMatches</a>&gt;&gt;&nbsp;getVariantMatches()</pre>
<div class="block">Returns the sequence variant matches of this peptide indexed by protein
 accession and peptide start.</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the sequence variant matches of this peptide</dd>
</dl>
</li>
</ul>
<a name="getMass--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getMass</h4>
<pre>public&nbsp;double&nbsp;getMass()</pre>
<div class="block">Getter for the mass.</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the peptide mass</dd>
</dl>
</li>
</ul>
<a name="getModificationMatches--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getModificationMatches</h4>
<pre>public&nbsp;<a href="../../../../../../com/compomics/util/experiment/identification/matches/ModificationMatch.html" title="class in com.compomics.util.experiment.identification.matches">ModificationMatch</a>[]&nbsp;getModificationMatches()</pre>
<div class="block">Getter for the modifications carried by this peptide.</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the modifications matches as found by the search engine</dd>
</dl>
</li>
</ul>
<a name="setModificationMatches-com.compomics.util.experiment.identification.matches.ModificationMatch:A-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>setModificationMatches</h4>
<pre>public&nbsp;void&nbsp;setModificationMatches(<a href="../../../../../../com/compomics/util/experiment/identification/matches/ModificationMatch.html" title="class in com.compomics.util.experiment.identification.matches">ModificationMatch</a>[]&nbsp;modificationMatches)</pre>
<div class="block">Sets new modification matches for the peptide.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>modificationMatches</code> - the new modification matches</dd>
</dl>
</li>
</ul>
<a name="clearModificationMatches--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>clearModificationMatches</h4>
<pre>public&nbsp;void&nbsp;clearModificationMatches()</pre>
<div class="block">Clears the list of imported modification matches.</div>
</li>
</ul>
<a name="addModificationMatch-com.compomics.util.experiment.identification.matches.ModificationMatch-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>addModificationMatch</h4>
<pre>public&nbsp;void&nbsp;addModificationMatch(<a href="../../../../../../com/compomics/util/experiment/identification/matches/ModificationMatch.html" title="class in com.compomics.util.experiment.identification.matches">ModificationMatch</a>&nbsp;modificationMatch)</pre>
<div class="block">Adds a modification match.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>modificationMatch</code> - the modification match to add</dd>
</dl>
</li>
</ul>
<a name="clearVariantMatches--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>clearVariantMatches</h4>
<pre>public&nbsp;void&nbsp;clearVariantMatches()</pre>
<div class="block">Clears the list of imported variant matches.</div>
</li>
</ul>
<a name="getPeptideEnd-java.lang.String-int-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getPeptideEnd</h4>
<pre>public&nbsp;int&nbsp;getPeptideEnd(<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;proteinAccession,
                         int&nbsp;peptideStart)</pre>
<div class="block">Returns the 0 based end index of the peptide on the protein sequence.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>proteinAccession</code> - the protein accession</dd>
<dd><code>peptideStart</code> - the peptide start index</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the 0 based end index of the peptide on the protein sequence</dd>
</dl>
</li>
</ul>
<a name="getSequence--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getSequence</h4>
<pre>public&nbsp;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;getSequence()</pre>
<div class="block">Returns for the sequence.</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the peptide sequence</dd>
</dl>
</li>
</ul>
<a name="setSequence-java.lang.String-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>setSequence</h4>
<pre>public&nbsp;void&nbsp;setSequence(<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;sequence)</pre>
<div class="block">Sets for the sequence.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>sequence</code> - the peptide sequence</dd>
</dl>
</li>
</ul>
<a name="getNMissedCleavages-com.compomics.util.experiment.biology.enzymes.Enzyme-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getNMissedCleavages</h4>
<pre>public&nbsp;int&nbsp;getNMissedCleavages(<a href="../../../../../../com/compomics/util/experiment/biology/enzymes/Enzyme.html" title="class in com.compomics.util.experiment.biology.enzymes">Enzyme</a>&nbsp;enzyme)</pre>
<div class="block">Returns the number of missed cleavages using the specified enzyme.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>enzyme</code> - the enzyme used</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the amount of missed cleavages</dd>
</dl>
</li>
</ul>
<a name="getNMissedCleavages-com.compomics.util.parameters.identification.search.DigestionParameters-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getNMissedCleavages</h4>
<pre>public&nbsp;int&nbsp;getNMissedCleavages(<a href="../../../../../../com/compomics/util/parameters/identification/search/DigestionParameters.html" title="class in com.compomics.util.parameters.identification.search">DigestionParameters</a>&nbsp;digestionPreferences)</pre>
<div class="block">Returns the number of missed cleavages using the digestion preferences. 0
 if no cleavage set. If multiple enzymes were used, the minimum across the
 different enzymes.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>digestionPreferences</code> - the digestion preferences</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the amount of missed cleavages</dd>
</dl>
</li>
</ul>
<a name="getMatchingKey--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getMatchingKey</h4>
<pre>public&nbsp;long&nbsp;getMatchingKey()</pre>
<div class="block">Returns the key accounting for sequence matching preferences</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the key accounting for sequence matching preferences</dd>
</dl>
</li>
</ul>
<a name="setMatchingKey-long-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>setMatchingKey</h4>
<pre>public&nbsp;void&nbsp;setMatchingKey(long&nbsp;matchingKey)</pre>
<div class="block">Sets the key accounting for sequence matching preferences.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>matchingKey</code> - the key accounting for sequence matching preferences</dd>
</dl>
</li>
</ul>
<a name="getMatchingKey-com.compomics.util.parameters.identification.advanced.SequenceMatchingParameters-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getMatchingKey</h4>
<pre>public&nbsp;long&nbsp;getMatchingKey(<a href="../../../../../../com/compomics/util/parameters/identification/advanced/SequenceMatchingParameters.html" title="class in com.compomics.util.parameters.identification.advanced">SequenceMatchingParameters</a>&nbsp;sequenceMatchingPreferences)</pre>
<div class="block">Returns a unique key for the peptide when considering the given matching
 preferences. When ambiguity the first amino acid according to
 AminoAcid.getAminoAcidsList() will be selected. For example the matching
 key of peptide PEPTLDE_mod1_mod2 is PEPTIDE_mod1_mod2</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>sequenceMatchingPreferences</code> - the sequence matching preferences</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>a key unique to the given matching type</dd>
</dl>
</li>
</ul>
<a name="getKey--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getKey</h4>
<pre>public&nbsp;long&nbsp;getKey()</pre>
<div class="block">Returns the reference key of a peptide. index =
 SEQUENCE_modMass1_modMass2 with modMass1 and modMass2 modification masses
 ordered alphabetically. See ExperimentObject for the conversion to long.

 Note: the key is not unique for indistinguishable sequences, see
 getMatchingKey(SequenceMatchingPreferences sequenceMatchingPreferences).
 Modifications must be loaded in the modification factory.</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the key of the peptide</dd>
</dl>
</li>
</ul>
<a name="getKey-java.lang.String-com.compomics.util.experiment.identification.matches.ModificationMatch:A-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getKey</h4>
<pre>public static&nbsp;long&nbsp;getKey(<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;sequence,
                          <a href="../../../../../../com/compomics/util/experiment/identification/matches/ModificationMatch.html" title="class in com.compomics.util.experiment.identification.matches">ModificationMatch</a>[]&nbsp;modificationMatches)</pre>
<div class="block">Returns the reference key of a peptide. key = SEQUENCE_mod1_mod2 modMass1
 and modMass2 modification masses ordered alphabetically.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>sequence</code> - the sequence of the peptide</dd>
<dd><code>modificationMatches</code> - list of modification matches</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the key of the peptide</dd>
</dl>
</li>
</ul>
<a name="getNVariableModifications-double-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getNVariableModifications</h4>
<pre>public&nbsp;int&nbsp;getNVariableModifications(double&nbsp;modificationMass)</pre>
<div class="block">Returns the number of variable modifications found with the given mass.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>modificationMass</code> - the mass of the modification</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the number of occurrences of this modification</dd>
</dl>
</li>
</ul>
<a name="getNModifications--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getNModifications</h4>
<pre>public&nbsp;int&nbsp;getNModifications()</pre>
<div class="block">Returns the number of modifications carried by this peptide.</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the number of modifications carried by this peptide</dd>
</dl>
</li>
</ul>
<a name="getPotentialModificationSitesNoCombination-com.compomics.util.experiment.biology.modifications.Modification-java.lang.String-int-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getPotentialModificationSitesNoCombination</h4>
<pre>public&nbsp;<a href="http://docs.oracle.com/javase/7/docs/api/java/util/ArrayList.html?is-external=true" title="class or interface in java.util">ArrayList</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Integer.html?is-external=true" title="class or interface in java.lang">Integer</a>&gt;&nbsp;getPotentialModificationSitesNoCombination(<a href="../../../../../../com/compomics/util/experiment/biology/modifications/Modification.html" title="class in com.compomics.util.experiment.biology.modifications">Modification</a>&nbsp;modification,
                                                                     <a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;proteinSequence,
                                                                     int&nbsp;peptideStart)</pre>
<div class="block">Returns the potential modification sites as an ordered list of sites. No
 amino acid combination is tested. 1 is the first amino acid. An empty
 list is returned if no possibility was found. No peptide to protein
 mapping is done. The index on the protein must be provided with 0 as
 first amino acid.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>modification</code> - the Modification considered</dd>
<dd><code>proteinSequence</code> - the protein sequence</dd>
<dd><code>peptideStart</code> - the index of the peptide start on the protein</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>a list of potential modification sites</dd>
</dl>
</li>
</ul>
<a name="getPotentialModificationSites-com.compomics.util.experiment.biology.modifications.Modification-com.compomics.util.experiment.io.biology.protein.SequenceProvider-com.compomics.util.parameters.identification.advanced.SequenceMatchingParameters-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getPotentialModificationSites</h4>
<pre>public&nbsp;<a href="http://docs.oracle.com/javase/7/docs/api/java/util/HashSet.html?is-external=true" title="class or interface in java.util">HashSet</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Integer.html?is-external=true" title="class or interface in java.lang">Integer</a>&gt;&nbsp;getPotentialModificationSites(<a href="../../../../../../com/compomics/util/experiment/biology/modifications/Modification.html" title="class in com.compomics.util.experiment.biology.modifications">Modification</a>&nbsp;modification,
                                                      <a href="../../../../../../com/compomics/util/experiment/io/biology/protein/SequenceProvider.html" title="interface in com.compomics.util.experiment.io.biology.protein">SequenceProvider</a>&nbsp;sequenceProvider,
                                                      <a href="../../../../../../com/compomics/util/parameters/identification/advanced/SequenceMatchingParameters.html" title="class in com.compomics.util.parameters.identification.advanced">SequenceMatchingParameters</a>&nbsp;modificationSequenceMatchingPreferences)</pre>
<div class="block">Returns the potential modification sites as a set. 1
 is the first amino acid. An empty list is returned if no modification
 site was found. All proteins and all positions are used.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>modification</code> - the Modification considered</dd>
<dd><code>sequenceProvider</code> - a protein sequence provider</dd>
<dd><code>modificationSequenceMatchingPreferences</code> - the sequence matching
 preferences for Modification to peptide mapping</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>a list of potential modification sites</dd>
</dl>
</li>
</ul>
<a name="getPotentialModificationSites-com.compomics.util.experiment.biology.modifications.Modification-java.lang.String-int-com.compomics.util.parameters.identification.advanced.SequenceMatchingParameters-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getPotentialModificationSites</h4>
<pre>public&nbsp;<a href="http://docs.oracle.com/javase/7/docs/api/java/util/ArrayList.html?is-external=true" title="class or interface in java.util">ArrayList</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Integer.html?is-external=true" title="class or interface in java.lang">Integer</a>&gt;&nbsp;getPotentialModificationSites(<a href="../../../../../../com/compomics/util/experiment/biology/modifications/Modification.html" title="class in com.compomics.util.experiment.biology.modifications">Modification</a>&nbsp;modification,
                                                        <a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;proteinSequence,
                                                        int&nbsp;peptideStart,
                                                        <a href="../../../../../../com/compomics/util/parameters/identification/advanced/SequenceMatchingParameters.html" title="class in com.compomics.util.parameters.identification.advanced">SequenceMatchingParameters</a>&nbsp;modificationSequenceMatchingPreferences)</pre>
<div class="block">Returns the potential modification sites as an ordered list of sites. 1
 is the first amino acid. An empty list is returned if no modification
 site was found.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>modification</code> - the Modification considered</dd>
<dd><code>proteinSequence</code> - the protein sequence</dd>
<dd><code>peptideStart</code> - the index of the peptide start on the protein</dd>
<dd><code>modificationSequenceMatchingPreferences</code> - the sequence matching
 preferences for Modification to peptide mapping</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>a list of potential modification sites</dd>
</dl>
</li>
</ul>
<a name="isNterm-java.lang.String-int-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>isNterm</h4>
<pre>public&nbsp;boolean&nbsp;isNterm(<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;proteinSequence,
                       int&nbsp;peptideStart)</pre>
<div class="block">Returns a boolean indicating whether the peptide is at the N-terminus of
 the given protein sequence. Initial methionine is cleaved.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>proteinSequence</code> - the sequence of the protein</dd>
<dd><code>peptideStart</code> - the 0 based peptide start index on the protein</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>a boolean indicating whether the peptide is at the N-terminus of
 the given protein</dd>
</dl>
</li>
</ul>
<a name="isCterm-java.lang.String-int-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>isCterm</h4>
<pre>public&nbsp;boolean&nbsp;isCterm(<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;proteinSequence,
                       int&nbsp;peptideStart)</pre>
<div class="block">Returns a boolean indicating whether the peptide is at the C-terminus of
 the given protein sequence.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>proteinSequence</code> - the sequence of the protein</dd>
<dd><code>peptideStart</code> - the 0 based peptide start index on the protein</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>a boolean indicating whether the peptide is at the C-terminus of
 the given protein</dd>
</dl>
</li>
</ul>
<a name="isSameSequenceAndModificationStatus-com.compomics.util.experiment.biology.proteins.Peptide-com.compomics.util.parameters.identification.advanced.SequenceMatchingParameters-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>isSameSequenceAndModificationStatus</h4>
<pre>public&nbsp;boolean&nbsp;isSameSequenceAndModificationStatus(<a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;anotherPeptide,
                                                   <a href="../../../../../../com/compomics/util/parameters/identification/advanced/SequenceMatchingParameters.html" title="class in com.compomics.util.parameters.identification.advanced">SequenceMatchingParameters</a>&nbsp;sequenceMatchingPreferences)</pre>
<div class="block">Indicates whether another peptide has the same sequence and modification
 status without accounting for modification localization.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>anotherPeptide</code> - the other peptide to compare to this instance</dd>
<dd><code>sequenceMatchingPreferences</code> - the sequence matching preferences</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>a boolean indicating whether the other peptide has the same
 sequence and modification status.</dd>
</dl>
</li>
</ul>
<a name="isSameSequence-com.compomics.util.experiment.biology.proteins.Peptide-com.compomics.util.parameters.identification.advanced.SequenceMatchingParameters-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>isSameSequence</h4>
<pre>public&nbsp;boolean&nbsp;isSameSequence(<a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;anotherPeptide,
                              <a href="../../../../../../com/compomics/util/parameters/identification/advanced/SequenceMatchingParameters.html" title="class in com.compomics.util.parameters.identification.advanced">SequenceMatchingParameters</a>&nbsp;sequenceMatchingPreferences)</pre>
<div class="block">Returns a boolean indicating whether another peptide has the same
 sequence as the given peptide</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>anotherPeptide</code> - the other peptide to compare</dd>
<dd><code>sequenceMatchingPreferences</code> - the sequence matching preferences</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>a boolean indicating whether the other peptide has the same
 sequence</dd>
</dl>
</li>
</ul>
<a name="isSameModificationStatus-com.compomics.util.experiment.biology.proteins.Peptide-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>isSameModificationStatus</h4>
<pre>public&nbsp;boolean&nbsp;isSameModificationStatus(<a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;anotherPeptide)</pre>
<div class="block">Indicates whether another peptide has the same variable modifications as
 this peptide. The localization of the Modification is not accounted for.
 Modifications are considered equal when of exact same mass, no rounding
 is conducted. Modifications should be loaded in the Modification factory.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>anotherPeptide</code> - the other peptide</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>a boolean indicating whether the other peptide has the same
 variable modifications as the peptide of interest</dd>
</dl>
</li>
</ul>
<a name="sameModificationsAs-com.compomics.util.experiment.biology.proteins.Peptide-java.util.ArrayList-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>sameModificationsAs</h4>
<pre>public&nbsp;boolean&nbsp;sameModificationsAs(<a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;anotherPeptide,
                                   <a href="http://docs.oracle.com/javase/7/docs/api/java/util/ArrayList.html?is-external=true" title="class or interface in java.util">ArrayList</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&gt;&nbsp;modifications)</pre>
<div class="block">Indicates whether another peptide has the same modifications at the same
 localization as this peptide. This method comes as a complement of
 isSameAs, here the localization of all Modifications is taken into
 account. Modifications are considered equal when of same mass.
 Modifications should be loaded in the Modification factory.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>anotherPeptide</code> - another peptide</dd>
<dd><code>modifications</code> - the Modifications</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>true if the other peptide has the same positions at the same
 location as the considered peptide</dd>
</dl>
</li>
</ul>
<a name="sameModificationsAs-com.compomics.util.experiment.biology.proteins.Peptide-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>sameModificationsAs</h4>
<pre>public&nbsp;boolean&nbsp;sameModificationsAs(<a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;anotherPeptide)</pre>
<div class="block">Indicates whether another peptide has the same modifications at the same
 localization as this peptide. This method comes as a complement of
 isSameAs, here the localization of all Modifications is taken into
 account. Modifications are considered equal when of same mass.
 Modifications should be loaded in the Modification factory.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>anotherPeptide</code> - another peptide</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>true if the other peptide has the same positions at the same
 location as the considered peptide</dd>
</dl>
</li>
</ul>
<a name="getNTerminal--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getNTerminal</h4>
<pre>public&nbsp;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;getNTerminal()</pre>
<div class="block">Returns the N-terminal of the peptide as a String. Returns "NH2" if the
 terminal is not modified, otherwise returns the name of the modification.
 /!\ this method will work only if the Modification found in the peptide
 are in the ModificationFactory.</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the N-terminal of the peptide as a String, e.g., "NH2"</dd>
</dl>
</li>
</ul>
<a name="getCTerminal--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getCTerminal</h4>
<pre>public&nbsp;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;getCTerminal()</pre>
<div class="block">Returns the C-terminal of the peptide as a String. Returns "COOH" if the
 terminal is not modified, otherwise returns the name of the modification.
 /!\ This method will work only if the Modification found in the peptide
 are in the ModificationFactory.</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the C-terminal of the peptide as a String, e.g., "COOH"</dd>
</dl>
</li>
</ul>
<a name="getTaggedModifiedSequence-com.compomics.util.parameters.identification.search.ModificationParameters-boolean-boolean-boolean-boolean-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getTaggedModifiedSequence</h4>
<pre>public&nbsp;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;getTaggedModifiedSequence(<a href="../../../../../../com/compomics/util/parameters/identification/search/ModificationParameters.html" title="class in com.compomics.util.parameters.identification.search">ModificationParameters</a>&nbsp;modificationProfile,
                                        boolean&nbsp;useHtmlColorCoding,
                                        boolean&nbsp;includeHtmlStartEndTags,
                                        boolean&nbsp;useShortName,
                                        boolean&nbsp;excludeAllFixedPtms)</pre>
<div class="block">Returns the modified sequence as an tagged string with potential
 modification sites color coded or with Modification tags, e.g,
 &lt;mox&gt;. /!\ this method will work only if the Modification found in
 the peptide are in the ModificationFactory. /!\ This method uses the
 modifications as set in the modification matches of this peptide and
 displays all of them.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>modificationProfile</code> - the modification profile of the search</dd>
<dd><code>useHtmlColorCoding</code> - if true, color coded HTML is used, otherwise
 Modification tags, e.g, &lt;mox&gt;, are used</dd>
<dd><code>includeHtmlStartEndTags</code> - if true, start and end HTML tags are added</dd>
<dd><code>useShortName</code> - if true the short names are used in the tags</dd>
<dd><code>excludeAllFixedPtms</code> - if true, all fixed Modifications are excluded</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the modified sequence as a tagged string</dd>
</dl>
</li>
</ul>
<a name="getTaggedModifiedSequence-com.compomics.util.parameters.identification.search.ModificationParameters-boolean-boolean-boolean-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getTaggedModifiedSequence</h4>
<pre>public&nbsp;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;getTaggedModifiedSequence(<a href="../../../../../../com/compomics/util/parameters/identification/search/ModificationParameters.html" title="class in com.compomics.util.parameters.identification.search">ModificationParameters</a>&nbsp;modificationProfile,
                                        boolean&nbsp;useHtmlColorCoding,
                                        boolean&nbsp;includeHtmlStartEndTags,
                                        boolean&nbsp;useShortName)</pre>
<div class="block">Returns the modified sequence as an tagged string with potential
 modification sites color coded or with Modification tags, e.g,
 &lt;mox&gt;. /!\ this method will work only if the Modification found in
 the peptide are in the ModificationFactory. /!\ This method uses the
 modifications as set in the modification matches of this peptide and
 displays all of them.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>modificationProfile</code> - the modification profile of the search</dd>
<dd><code>useHtmlColorCoding</code> - if true, color coded HTML is used, otherwise
 Modification tags, e.g, &lt;mox&gt;, are used</dd>
<dd><code>includeHtmlStartEndTags</code> - if true, start and end HTML tags are added</dd>
<dd><code>useShortName</code> - if true the short names are used in the tags</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the modified sequence as a tagged string</dd>
</dl>
</li>
</ul>
<a name="estimateTheoreticMass--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>estimateTheoreticMass</h4>
<pre>public&nbsp;void&nbsp;estimateTheoreticMass()</pre>
<div class="block">Estimates the theoretic mass of the peptide. The previous version is
 silently overwritten.</div>
</li>
</ul>
<a name="getNoModPeptide-java.util.HashSet-">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>getNoModPeptide</h4>
<pre>public&nbsp;<a href="../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;getNoModPeptide(<a href="http://docs.oracle.com/javase/7/docs/api/java/util/HashSet.html?is-external=true" title="class or interface in java.util">HashSet</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&gt;&nbsp;forbiddenModifications)</pre>
<div class="block">Returns a version of the peptide which does not contain the given list of
 modifications.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>forbiddenModifications</code> - list of forbidden modifications</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>a not modified version of the peptide</dd>
</dl>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
</div>
<!-- ========= END OF CLASS DATA ========= -->
<!-- ======= START OF BOTTOM NAVBAR ====== -->
<div class="bottomNav"><a name="navbar.bottom">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.bottom" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.bottom.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../../../../../../overview-summary.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="navBarCell1Rev">Class</li>
<li><a href="class-use/Peptide.html">Use</a></li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../../../../../../deprecated-list.html">Deprecated</a></li>
<li><a href="../../../../../../index-all.html">Index</a></li>
<li><a href="../../../../../../help-doc.html">Help</a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li>Prev&nbsp;Class</li>
<li><a href="../../../../../../com/compomics/util/experiment/biology/proteins/Protein.html" title="class in com.compomics.util.experiment.biology.proteins"><span class="typeNameLink">Next&nbsp;Class</span></a></li>
</ul>
<ul class="navList">
<li><a href="../../../../../../index.html?com/compomics/util/experiment/biology/proteins/Peptide.html" target="_top">Frames</a></li>
<li><a href="Peptide.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_bottom">
<li><a href="../../../../../../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_bottom");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<div>
<ul class="subNavList">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li><a href="#field.summary">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.summary">Method</a></li>
</ul>
<ul class="subNavList">
<li>Detail:&nbsp;</li>
<li><a href="#field.detail">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.detail">Method</a></li>
</ul>
</div>
<a name="skip.navbar.bottom">
<!--   -->
</a></div>
<!-- ======== END OF BOTTOM NAVBAR ======= -->
<p class="legalCopy"><small>Copyright &#169; 2018. All rights reserved.</small></p>
</body>
</html>
