<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!-- NewPage -->
<html lang="en">
<head>
<!-- Generated by javadoc -->
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<title>Uses of Class com.compomics.util.experiment.biology.proteins.Peptide (computational omics utilities 5.0.1-SNAPSHOT API)</title>
<link rel="stylesheet" type="text/css" href="../../../../../../../stylesheet.css" title="Style">
<script type="text/javascript" src="../../../../../../../script.js"></script>
</head>
<body>
<script type="text/javascript"><!--
    try {
        if (location.href.indexOf('is-external=true') == -1) {
            parent.document.title="Uses of Class com.compomics.util.experiment.biology.proteins.Peptide (computational omics utilities 5.0.1-SNAPSHOT API)";
        }
    }
    catch(err) {
    }
//-->
</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
<!-- ========= START OF TOP NAVBAR ======= -->
<div class="topNav"><a name="navbar.top">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.top" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.top.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../../../../../../../overview-summary.html">Overview</a></li>
<li><a href="../package-summary.html">Package</a></li>
<li><a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Class</a></li>
<li class="navBarCell1Rev">Use</li>
<li><a href="../package-tree.html">Tree</a></li>
<li><a href="../../../../../../../deprecated-list.html">Deprecated</a></li>
<li><a href="../../../../../../../index-all.html">Index</a></li>
<li><a href="../../../../../../../help-doc.html">Help</a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li>Prev</li>
<li>Next</li>
</ul>
<ul class="navList">
<li><a href="../../../../../../../index.html?com/compomics/util/experiment/biology/proteins/class-use/Peptide.html" target="_top">Frames</a></li>
<li><a href="Peptide.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_top">
<li><a href="../../../../../../../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_top");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<a name="skip.navbar.top">
<!--   -->
</a></div>
<!-- ========= END OF TOP NAVBAR ========= -->
<div class="header">
<h2 title="Uses of Class com.compomics.util.experiment.biology.proteins.Peptide" class="title">Uses of Class<br>com.compomics.util.experiment.biology.proteins.Peptide</h2>
</div>
<div class="classUseContainer">
<ul class="blockList">
<li class="blockList">
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing packages, and an explanation">
<caption><span>Packages that use <a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Package</th>
<th class="colLast" scope="col">Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><a href="#com.compomics.util.experiment.biology.ions">com.compomics.util.experiment.biology.ions</a></td>
<td class="colLast">
<div class="block">Experiment classes related to ions.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><a href="#com.compomics.util.experiment.biology.proteins">com.compomics.util.experiment.biology.proteins</a></td>
<td class="colLast">
<div class="block">Classes related to proteins.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><a href="#com.compomics.util.experiment.identification">com.compomics.util.experiment.identification</a></td>
<td class="colLast">
<div class="block">Experiment classes related to identifications.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><a href="#com.compomics.util.experiment.identification.filtering">com.compomics.util.experiment.identification.filtering</a></td>
<td class="colLast">
<div class="block">Classes for the filtering of identification objects.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><a href="#com.compomics.util.experiment.identification.matches">com.compomics.util.experiment.identification.matches</a></td>
<td class="colLast">
<div class="block">Experiment classes related to matches.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><a href="#com.compomics.util.experiment.identification.modification">com.compomics.util.experiment.identification.modification</a></td>
<td class="colLast">
<div class="block">Additional PSM scoring classes.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><a href="#com.compomics.util.experiment.identification.modification.scores">com.compomics.util.experiment.identification.modification.scores</a></td>
<td class="colLast">
<div class="block">PTM scores.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><a href="#com.compomics.util.experiment.identification.peptide_fragmentation.models.ms2pip.features_generation">com.compomics.util.experiment.identification.peptide_fragmentation.models.ms2pip.features_generation</a></td>
<td class="colLast">
<div class="block">This package contains the classes needed to generate ms2pip features.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><a href="#com.compomics.util.experiment.identification.protein_inference">com.compomics.util.experiment.identification.protein_inference</a></td>
<td class="colLast">
<div class="block">Classes related to protein inference.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><a href="#com.compomics.util.experiment.identification.protein_sequences.digestion">com.compomics.util.experiment.identification.protein_sequences.digestion</a></td>
<td class="colLast">
<div class="block">This package contains classes used to digest protein sequences into peptides.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><a href="#com.compomics.util.experiment.identification.psm_scoring">com.compomics.util.experiment.identification.psm_scoring</a></td>
<td class="colLast">
<div class="block">Main PSM scoring class.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><a href="#com.compomics.util.experiment.identification.psm_scoring.psm_scores">com.compomics.util.experiment.identification.psm_scoring.psm_scores</a></td>
<td class="colLast">
<div class="block">Main PTM scoring classes.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><a href="#com.compomics.util.experiment.identification.spectrum_annotation.simple_annotators">com.compomics.util.experiment.identification.spectrum_annotation.simple_annotators</a></td>
<td class="colLast">
<div class="block">This package contains implementations of the SimpleAnnotator allowing the annotation of specific sets of ions.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><a href="#com.compomics.util.experiment.identification.spectrum_annotation.spectrum_annotators">com.compomics.util.experiment.identification.spectrum_annotation.spectrum_annotators</a></td>
<td class="colLast">
<div class="block">This package contains implementations of the SpectrumAnnotator allowing the annotation of all ions.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><a href="#com.compomics.util.experiment.identification.spectrum_assumptions">com.compomics.util.experiment.identification.spectrum_assumptions</a></td>
<td class="colLast">
<div class="block">Implementations of the SpectrumIdentificationAssumption.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><a href="#com.compomics.util.experiment.identification.utils">com.compomics.util.experiment.identification.utils</a></td>
<td class="colLast">&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><a href="#com.compomics.util.gui.spectrum">com.compomics.util.gui.spectrum</a></td>
<td class="colLast">
<div class="block">Spectrum and Chromatogram visualization GUI classes.</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList">
<ul class="blockList">
<li class="blockList"><a name="com.compomics.util.experiment.biology.ions">
<!--   -->
</a>
<h3>Uses of <a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a> in <a href="../../../../../../../com/compomics/util/experiment/biology/ions/package-summary.html">com.compomics.util.experiment.biology.ions</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing methods, and an explanation">
<caption><span>Methods in <a href="../../../../../../../com/compomics/util/experiment/biology/ions/package-summary.html">com.compomics.util.experiment.biology.ions</a> with parameters of type <a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code><a href="http://docs.oracle.com/javase/7/docs/api/java/util/HashMap.html?is-external=true" title="class or interface in java.util">HashMap</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Integer.html?is-external=true" title="class or interface in java.lang">Integer</a>,<a href="http://docs.oracle.com/javase/7/docs/api/java/util/HashMap.html?is-external=true" title="class or interface in java.util">HashMap</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Integer.html?is-external=true" title="class or interface in java.lang">Integer</a>,<a href="http://docs.oracle.com/javase/7/docs/api/java/util/ArrayList.html?is-external=true" title="class or interface in java.util">ArrayList</a>&lt;<a href="../../../../../../../com/compomics/util/experiment/biology/ions/Ion.html" title="class in com.compomics.util.experiment.biology.ions">Ion</a>&gt;&gt;&gt;</code></td>
<td class="colLast"><span class="typeNameLabel">IonFactory.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/biology/ions/IonFactory.html#getFragmentIons-com.compomics.util.experiment.biology.proteins.Peptide-">getFragmentIons</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;peptide)</code>
<div class="block">This method returns all the theoretic ions expected from a peptide.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code><a href="http://docs.oracle.com/javase/7/docs/api/java/util/HashMap.html?is-external=true" title="class or interface in java.util">HashMap</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Integer.html?is-external=true" title="class or interface in java.lang">Integer</a>,<a href="http://docs.oracle.com/javase/7/docs/api/java/util/HashMap.html?is-external=true" title="class or interface in java.util">HashMap</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Integer.html?is-external=true" title="class or interface in java.lang">Integer</a>,<a href="http://docs.oracle.com/javase/7/docs/api/java/util/ArrayList.html?is-external=true" title="class or interface in java.util">ArrayList</a>&lt;<a href="../../../../../../../com/compomics/util/experiment/biology/ions/Ion.html" title="class in com.compomics.util.experiment.biology.ions">Ion</a>&gt;&gt;&gt;</code></td>
<td class="colLast"><span class="typeNameLabel">IonFactory.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/biology/ions/IonFactory.html#getFragmentIons-com.compomics.util.experiment.biology.proteins.Peptide-com.compomics.util.experiment.identification.spectrum_annotation.SpecificAnnotationParameters-">getFragmentIons</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;peptide,
               <a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/SpecificAnnotationParameters.html" title="class in com.compomics.util.experiment.identification.spectrum_annotation">SpecificAnnotationParameters</a>&nbsp;specificAnnotationSettings)</code>
<div class="block">This method returns the theoretic ions expected from a peptide.</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="com.compomics.util.experiment.biology.proteins">
<!--   -->
</a>
<h3>Uses of <a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a> in <a href="../../../../../../../com/compomics/util/experiment/biology/proteins/package-summary.html">com.compomics.util.experiment.biology.proteins</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing methods, and an explanation">
<caption><span>Methods in <a href="../../../../../../../com/compomics/util/experiment/biology/proteins/package-summary.html">com.compomics.util.experiment.biology.proteins</a> that return <a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code><a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a></code></td>
<td class="colLast"><span class="typeNameLabel">Peptide.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html#getNoModPeptide-java.util.HashSet-">getNoModPeptide</a></span>(<a href="http://docs.oracle.com/javase/7/docs/api/java/util/HashSet.html?is-external=true" title="class or interface in java.util">HashSet</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&gt;&nbsp;forbiddenModifications)</code>
<div class="block">Returns a version of the peptide which does not contain the given list of
 modifications.</div>
</td>
</tr>
</tbody>
</table>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing methods, and an explanation">
<caption><span>Methods in <a href="../../../../../../../com/compomics/util/experiment/biology/proteins/package-summary.html">com.compomics.util.experiment.biology.proteins</a> with parameters of type <a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><span class="typeNameLabel">Peptide.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html#isSameModificationStatus-com.compomics.util.experiment.biology.proteins.Peptide-">isSameModificationStatus</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;anotherPeptide)</code>
<div class="block">Indicates whether another peptide has the same variable modifications as
 this peptide.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><span class="typeNameLabel">Peptide.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html#isSameSequence-com.compomics.util.experiment.biology.proteins.Peptide-com.compomics.util.parameters.identification.advanced.SequenceMatchingParameters-">isSameSequence</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;anotherPeptide,
              <a href="../../../../../../../com/compomics/util/parameters/identification/advanced/SequenceMatchingParameters.html" title="class in com.compomics.util.parameters.identification.advanced">SequenceMatchingParameters</a>&nbsp;sequenceMatchingPreferences)</code>
<div class="block">Returns a boolean indicating whether another peptide has the same
 sequence as the given peptide</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><span class="typeNameLabel">Peptide.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html#isSameSequenceAndModificationStatus-com.compomics.util.experiment.biology.proteins.Peptide-com.compomics.util.parameters.identification.advanced.SequenceMatchingParameters-">isSameSequenceAndModificationStatus</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;anotherPeptide,
                                   <a href="../../../../../../../com/compomics/util/parameters/identification/advanced/SequenceMatchingParameters.html" title="class in com.compomics.util.parameters.identification.advanced">SequenceMatchingParameters</a>&nbsp;sequenceMatchingPreferences)</code>
<div class="block">Indicates whether another peptide has the same sequence and modification
 status without accounting for modification localization.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><span class="typeNameLabel">Peptide.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html#sameModificationsAs-com.compomics.util.experiment.biology.proteins.Peptide-">sameModificationsAs</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;anotherPeptide)</code>
<div class="block">Indicates whether another peptide has the same modifications at the same
 localization as this peptide.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><span class="typeNameLabel">Peptide.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html#sameModificationsAs-com.compomics.util.experiment.biology.proteins.Peptide-java.util.ArrayList-">sameModificationsAs</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;anotherPeptide,
                   <a href="http://docs.oracle.com/javase/7/docs/api/java/util/ArrayList.html?is-external=true" title="class or interface in java.util">ArrayList</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&gt;&nbsp;modifications)</code>
<div class="block">Indicates whether another peptide has the same modifications at the same
 localization as this peptide.</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="com.compomics.util.experiment.identification">
<!--   -->
</a>
<h3>Uses of <a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a> in <a href="../../../../../../../com/compomics/util/experiment/identification/package-summary.html">com.compomics.util.experiment.identification</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing methods, and an explanation">
<caption><span>Methods in <a href="../../../../../../../com/compomics/util/experiment/identification/package-summary.html">com.compomics.util.experiment.identification</a> with parameters of type <a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code><a href="http://docs.oracle.com/javase/7/docs/api/java/util/HashSet.html?is-external=true" title="class or interface in java.util">HashSet</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Long.html?is-external=true" title="class or interface in java.lang">Long</a>&gt;</code></td>
<td class="colLast"><span class="typeNameLabel">Identification.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/Identification.html#getProteinMatches-com.compomics.util.experiment.biology.proteins.Peptide-">getProteinMatches</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;peptide)</code>
<div class="block">Returns the keys of the protein matches where a peptide can be found.</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="com.compomics.util.experiment.identification.filtering">
<!--   -->
</a>
<h3>Uses of <a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a> in <a href="../../../../../../../com/compomics/util/experiment/identification/filtering/package-summary.html">com.compomics.util.experiment.identification.filtering</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing methods, and an explanation">
<caption><span>Methods in <a href="../../../../../../../com/compomics/util/experiment/identification/filtering/package-summary.html">com.compomics.util.experiment.identification.filtering</a> with parameters of type <a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><span class="typeNameLabel">PeptideAssumptionFilter.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/filtering/PeptideAssumptionFilter.html#validateModifications-com.compomics.util.experiment.biology.proteins.Peptide-com.compomics.util.parameters.identification.advanced.SequenceMatchingParameters-com.compomics.util.parameters.identification.advanced.SequenceMatchingParameters-com.compomics.util.parameters.identification.search.ModificationParameters-">validateModifications</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;peptide,
                     <a href="../../../../../../../com/compomics/util/parameters/identification/advanced/SequenceMatchingParameters.html" title="class in com.compomics.util.parameters.identification.advanced">SequenceMatchingParameters</a>&nbsp;sequenceMatchingPreferences,
                     <a href="../../../../../../../com/compomics/util/parameters/identification/advanced/SequenceMatchingParameters.html" title="class in com.compomics.util.parameters.identification.advanced">SequenceMatchingParameters</a>&nbsp;modificationSequenceMatchingPreferences,
                     <a href="../../../../../../../com/compomics/util/parameters/identification/search/ModificationParameters.html" title="class in com.compomics.util.parameters.identification.search">ModificationParameters</a>&nbsp;modificationProfile)</code>
<div class="block">Verifies that the definition of every modification name is available.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><span class="typeNameLabel">PeptideAssumptionFilter.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/filtering/PeptideAssumptionFilter.html#validatePeptide-com.compomics.util.experiment.biology.proteins.Peptide-com.compomics.util.parameters.identification.advanced.SequenceMatchingParameters-com.compomics.util.parameters.identification.search.DigestionParameters-">validatePeptide</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;peptide,
               <a href="../../../../../../../com/compomics/util/parameters/identification/advanced/SequenceMatchingParameters.html" title="class in com.compomics.util.parameters.identification.advanced">SequenceMatchingParameters</a>&nbsp;sequenceMatchingPreferences,
               <a href="../../../../../../../com/compomics/util/parameters/identification/search/DigestionParameters.html" title="class in com.compomics.util.parameters.identification.search">DigestionParameters</a>&nbsp;digestionPreferences)</code>
<div class="block">Validates the peptide based on the peptide length, the share of X's in
 the sequence and the allowed number of missed cleavages.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><span class="typeNameLabel">PeptideAssumptionFilter.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/filtering/PeptideAssumptionFilter.html#validateProteins-com.compomics.util.experiment.biology.proteins.Peptide-com.compomics.util.parameters.identification.advanced.SequenceMatchingParameters-com.compomics.util.experiment.io.biology.protein.SequenceProvider-">validateProteins</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;peptide,
                <a href="../../../../../../../com/compomics/util/parameters/identification/advanced/SequenceMatchingParameters.html" title="class in com.compomics.util.parameters.identification.advanced">SequenceMatchingParameters</a>&nbsp;sequenceMatchingPreferences,
                <a href="../../../../../../../com/compomics/util/experiment/io/biology/protein/SequenceProvider.html" title="interface in com.compomics.util.experiment.io.biology.protein">SequenceProvider</a>&nbsp;sequenceProvider)</code>
<div class="block">Validates a peptide depending on its protein inference status.</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="com.compomics.util.experiment.identification.matches">
<!--   -->
</a>
<h3>Uses of <a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a> in <a href="../../../../../../../com/compomics/util/experiment/identification/matches/package-summary.html">com.compomics.util.experiment.identification.matches</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing methods, and an explanation">
<caption><span>Methods in <a href="../../../../../../../com/compomics/util/experiment/identification/matches/package-summary.html">com.compomics.util.experiment.identification.matches</a> that return <a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code><a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a></code></td>
<td class="colLast"><span class="typeNameLabel">PeptideMatch.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/matches/PeptideMatch.html#getPeptide--">getPeptide</a></span>()</code>
<div class="block">Getter for the peptide.</div>
</td>
</tr>
</tbody>
</table>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing methods, and an explanation">
<caption><span>Methods in <a href="../../../../../../../com/compomics/util/experiment/identification/matches/package-summary.html">com.compomics.util.experiment.identification.matches</a> with parameters of type <a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>static long</code></td>
<td class="colLast"><span class="typeNameLabel">ProteinMatch.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/matches/ProteinMatch.html#getProteinMatchKey-com.compomics.util.experiment.biology.proteins.Peptide-">getProteinMatchKey</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;peptide)</code>
<div class="block">Convenience method which returns the protein key from a peptide.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><span class="typeNameLabel">PeptideMatch.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/matches/PeptideMatch.html#setPeptide-com.compomics.util.experiment.biology.proteins.Peptide-">setPeptide</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;peptide)</code>
<div class="block">Setter for the peptide.</div>
</td>
</tr>
</tbody>
</table>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing constructors, and an explanation">
<caption><span>Constructors in <a href="../../../../../../../com/compomics/util/experiment/identification/matches/package-summary.html">com.compomics.util.experiment.identification.matches</a> with parameters of type <a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colOne" scope="col">Constructor and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/matches/PeptideMatch.html#PeptideMatch-com.compomics.util.experiment.biology.proteins.Peptide-long-long-">PeptideMatch</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;peptide,
            long&nbsp;matchKey,
            long&nbsp;spectrumMatchKey)</code>
<div class="block">Constructor for the peptide match.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/matches/ProteinMatch.html#ProteinMatch-com.compomics.util.experiment.biology.proteins.Peptide-long-">ProteinMatch</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;peptide,
            long&nbsp;peptideMatchKey)</code>
<div class="block">Constructor for the protein match.</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="com.compomics.util.experiment.identification.modification">
<!--   -->
</a>
<h3>Uses of <a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a> in <a href="../../../../../../../com/compomics/util/experiment/identification/modification/package-summary.html">com.compomics.util.experiment.identification.modification</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing methods, and an explanation">
<caption><span>Methods in <a href="../../../../../../../com/compomics/util/experiment/identification/modification/package-summary.html">com.compomics.util.experiment.identification.modification</a> with parameters of type <a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>static <a href="http://docs.oracle.com/javase/7/docs/api/java/util/HashMap.html?is-external=true" title="class or interface in java.util">HashMap</a>&lt;<a href="../../../../../../../com/compomics/util/experiment/biology/ions/impl/PeptideFragmentIon.html" title="class in com.compomics.util.experiment.biology.ions.impl">PeptideFragmentIon</a>,<a href="http://docs.oracle.com/javase/7/docs/api/java/util/ArrayList.html?is-external=true" title="class or interface in java.util">ArrayList</a>&lt;<a href="../../../../../../../com/compomics/util/experiment/identification/matches/IonMatch.html" title="class in com.compomics.util.experiment.identification.matches">IonMatch</a>&gt;&gt;</code></td>
<td class="colLast"><span class="typeNameLabel">ModificationtableContent.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/modification/ModificationtableContent.html#getModificationPlotData-com.compomics.util.experiment.biology.proteins.Peptide-com.compomics.util.experiment.biology.modifications.Modification-int-com.compomics.util.experiment.mass_spectrometry.spectra.Spectrum-com.compomics.util.experiment.identification.spectrum_annotation.AnnotationParameters-com.compomics.util.experiment.identification.spectrum_annotation.SpecificAnnotationParameters-">getModificationPlotData</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;peptide,
                       <a href="../../../../../../../com/compomics/util/experiment/biology/modifications/Modification.html" title="class in com.compomics.util.experiment.biology.modifications">Modification</a>&nbsp;modification,
                       int&nbsp;nMod,
                       <a href="../../../../../../../com/compomics/util/experiment/mass_spectrometry/spectra/Spectrum.html" title="class in com.compomics.util.experiment.mass_spectrometry.spectra">Spectrum</a>&nbsp;spectrum,
                       <a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/AnnotationParameters.html" title="class in com.compomics.util.experiment.identification.spectrum_annotation">AnnotationParameters</a>&nbsp;annotationPreferences,
                       <a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/SpecificAnnotationParameters.html" title="class in com.compomics.util.experiment.identification.spectrum_annotation">SpecificAnnotationParameters</a>&nbsp;specificAnnotationPreferences)</code>
<div class="block">Returns the modification plot series in the JFreechart format for one
 PSM.</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="com.compomics.util.experiment.identification.modification.scores">
<!--   -->
</a>
<h3>Uses of <a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a> in <a href="../../../../../../../com/compomics/util/experiment/identification/modification/scores/package-summary.html">com.compomics.util.experiment.identification.modification.scores</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing methods, and an explanation">
<caption><span>Methods in <a href="../../../../../../../com/compomics/util/experiment/identification/modification/scores/package-summary.html">com.compomics.util.experiment.identification.modification.scores</a> with parameters of type <a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>static <a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Double.html?is-external=true" title="class or interface in java.lang">Double</a></code></td>
<td class="colLast"><span class="typeNameLabel">MDScore.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/modification/scores/MDScore.html#getMDScore-java.util.ArrayList-com.compomics.util.experiment.biology.proteins.Peptide-java.util.ArrayList-com.compomics.util.parameters.identification.advanced.SequenceMatchingParameters-java.lang.Integer-">getMDScore</a></span>(<a href="http://docs.oracle.com/javase/7/docs/api/java/util/ArrayList.html?is-external=true" title="class or interface in java.util">ArrayList</a>&lt;<a href="../../../../../../../com/compomics/util/experiment/identification/SpectrumIdentificationAssumption.html" title="class in com.compomics.util.experiment.identification">SpectrumIdentificationAssumption</a>&gt;&nbsp;mascotAssumptions,
          <a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;peptideCandidate,
          <a href="http://docs.oracle.com/javase/7/docs/api/java/util/ArrayList.html?is-external=true" title="class or interface in java.util">ArrayList</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&gt;&nbsp;ptms,
          <a href="../../../../../../../com/compomics/util/parameters/identification/advanced/SequenceMatchingParameters.html" title="class in com.compomics.util.parameters.identification.advanced">SequenceMatchingParameters</a>&nbsp;sequenceMatchingPreferences,
          <a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Integer.html?is-external=true" title="class or interface in java.lang">Integer</a>&nbsp;rounding)</code>
<div class="block">Returns the MD score for the given peptide in a spectrum match.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static <a href="http://docs.oracle.com/javase/7/docs/api/java/util/HashMap.html?is-external=true" title="class or interface in java.util">HashMap</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Integer.html?is-external=true" title="class or interface in java.lang">Integer</a>,<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Double.html?is-external=true" title="class or interface in java.lang">Double</a>&gt;</code></td>
<td class="colLast"><span class="typeNameLabel">PhosphoRS.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/modification/scores/PhosphoRS.html#getSequenceProbabilities-com.compomics.util.experiment.biology.proteins.Peptide-java.util.ArrayList-com.compomics.util.experiment.mass_spectrometry.spectra.Spectrum-com.compomics.util.experiment.io.biology.protein.SequenceProvider-com.compomics.util.experiment.identification.spectrum_annotation.AnnotationParameters-com.compomics.util.experiment.identification.spectrum_annotation.SpecificAnnotationParameters-boolean-com.compomics.util.parameters.identification.advanced.SequenceMatchingParameters-com.compomics.util.parameters.identification.advanced.SequenceMatchingParameters-com.compomics.util.experiment.identification.spectrum_annotation.spectrum_annotators.PeptideSpectrumAnnotator-">getSequenceProbabilities</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;peptide,
                        <a href="http://docs.oracle.com/javase/7/docs/api/java/util/ArrayList.html?is-external=true" title="class or interface in java.util">ArrayList</a>&lt;<a href="../../../../../../../com/compomics/util/experiment/biology/modifications/Modification.html" title="class in com.compomics.util.experiment.biology.modifications">Modification</a>&gt;&nbsp;modifications,
                        <a href="../../../../../../../com/compomics/util/experiment/mass_spectrometry/spectra/Spectrum.html" title="class in com.compomics.util.experiment.mass_spectrometry.spectra">Spectrum</a>&nbsp;spectrum,
                        <a href="../../../../../../../com/compomics/util/experiment/io/biology/protein/SequenceProvider.html" title="interface in com.compomics.util.experiment.io.biology.protein">SequenceProvider</a>&nbsp;sequenceProvider,
                        <a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/AnnotationParameters.html" title="class in com.compomics.util.experiment.identification.spectrum_annotation">AnnotationParameters</a>&nbsp;annotationSettings,
                        <a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/SpecificAnnotationParameters.html" title="class in com.compomics.util.experiment.identification.spectrum_annotation">SpecificAnnotationParameters</a>&nbsp;specificAnnotationSettings,
                        boolean&nbsp;accountNeutralLosses,
                        <a href="../../../../../../../com/compomics/util/parameters/identification/advanced/SequenceMatchingParameters.html" title="class in com.compomics.util.parameters.identification.advanced">SequenceMatchingParameters</a>&nbsp;sequenceMatchingPreferences,
                        <a href="../../../../../../../com/compomics/util/parameters/identification/advanced/SequenceMatchingParameters.html" title="class in com.compomics.util.parameters.identification.advanced">SequenceMatchingParameters</a>&nbsp;modificationSequenceMatchingPreferences,
                        <a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/spectrum_annotators/PeptideSpectrumAnnotator.html" title="class in com.compomics.util.experiment.identification.spectrum_annotation.spectrum_annotators">PeptideSpectrumAnnotator</a>&nbsp;spectrumAnnotator)</code>
<div class="block">Returns the PhosphoRS sequence probabilities for the modification
 possible locations.</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="com.compomics.util.experiment.identification.peptide_fragmentation.models.ms2pip.features_generation">
<!--   -->
</a>
<h3>Uses of <a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a> in <a href="../../../../../../../com/compomics/util/experiment/identification/peptide_fragmentation/models/ms2pip/features_generation/package-summary.html">com.compomics.util.experiment.identification.peptide_fragmentation.models.ms2pip.features_generation</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing methods, and an explanation">
<caption><span>Methods in <a href="../../../../../../../com/compomics/util/experiment/identification/peptide_fragmentation/models/ms2pip/features_generation/package-summary.html">com.compomics.util.experiment.identification.peptide_fragmentation.models.ms2pip.features_generation</a> with parameters of type <a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>int[]</code></td>
<td class="colLast"><span class="typeNameLabel">FeaturesGenerator.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/peptide_fragmentation/models/ms2pip/features_generation/FeaturesGenerator.html#getComplementaryIonsFeatures-com.compomics.util.experiment.biology.proteins.Peptide-int-int-">getComplementaryIonsFeatures</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;peptide,
                            int&nbsp;charge,
                            int&nbsp;ionIndex)</code>
<div class="block">Returns the ms2pip features for the complementary ions of the given
 peptide at the given charge.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>int[]</code></td>
<td class="colLast"><span class="typeNameLabel">FeaturesGenerator.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/peptide_fragmentation/models/ms2pip/features_generation/FeaturesGenerator.html#getForwardIonsFeatures-com.compomics.util.experiment.biology.proteins.Peptide-int-int-">getForwardIonsFeatures</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;peptide,
                      int&nbsp;charge,
                      int&nbsp;ionIndex)</code>
<div class="block">Returns the ms2pip features for the forward ions of the given peptide at
 the given charge.</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="com.compomics.util.experiment.identification.protein_inference">
<!--   -->
</a>
<h3>Uses of <a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a> in <a href="../../../../../../../com/compomics/util/experiment/identification/protein_inference/package-summary.html">com.compomics.util.experiment.identification.protein_inference</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing methods, and an explanation">
<caption><span>Methods in <a href="../../../../../../../com/compomics/util/experiment/identification/protein_inference/package-summary.html">com.compomics.util.experiment.identification.protein_inference</a> that return types with arguments of type <a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>static <a href="http://docs.oracle.com/javase/7/docs/api/java/util/Collection.html?is-external=true" title="class or interface in java.util">Collection</a>&lt;<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&gt;</code></td>
<td class="colLast"><span class="typeNameLabel">PeptideProteinMapping.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/protein_inference/PeptideProteinMapping.html#getPeptides-java.util.ArrayList-com.compomics.util.parameters.identification.advanced.SequenceMatchingParameters-">getPeptides</a></span>(<a href="http://docs.oracle.com/javase/7/docs/api/java/util/ArrayList.html?is-external=true" title="class or interface in java.util">ArrayList</a>&lt;<a href="../../../../../../../com/compomics/util/experiment/identification/protein_inference/PeptideProteinMapping.html" title="class in com.compomics.util.experiment.identification.protein_inference">PeptideProteinMapping</a>&gt;&nbsp;peptideProteinMappings,
           <a href="../../../../../../../com/compomics/util/parameters/identification/advanced/SequenceMatchingParameters.html" title="class in com.compomics.util.parameters.identification.advanced">SequenceMatchingParameters</a>&nbsp;sequenceMatchingPreferences)</code>
<div class="block">Aggregates the given mapping into a list of peptides.</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="com.compomics.util.experiment.identification.protein_sequences.digestion">
<!--   -->
</a>
<h3>Uses of <a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a> in <a href="../../../../../../../com/compomics/util/experiment/identification/protein_sequences/digestion/package-summary.html">com.compomics.util.experiment.identification.protein_sequences.digestion</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing methods, and an explanation">
<caption><span>Methods in <a href="../../../../../../../com/compomics/util/experiment/identification/protein_sequences/digestion/package-summary.html">com.compomics.util.experiment.identification.protein_sequences.digestion</a> that return <a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code><a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a></code></td>
<td class="colLast"><span class="typeNameLabel">PeptideWithPosition.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/protein_sequences/digestion/PeptideWithPosition.html#getPeptide--">getPeptide</a></span>()</code>
<div class="block">Returns the peptide.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code><a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a></code></td>
<td class="colLast"><span class="typeNameLabel">PeptideDraft.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/protein_sequences/digestion/PeptideDraft.html#getPeptide-double-double-">getPeptide</a></span>(double&nbsp;massMin,
          double&nbsp;massMax)</code>
<div class="block">Returns a peptide from the peptide draft.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code><a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a></code></td>
<td class="colLast"><span class="typeNameLabel">PeptideDraft.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/protein_sequences/digestion/PeptideDraft.html#getPeptide-double-double-com.compomics.util.general.BoxedObject-">getPeptide</a></span>(double&nbsp;massMin,
          double&nbsp;massMax,
          <a href="../../../../../../../com/compomics/util/general/BoxedObject.html" title="class in com.compomics.util.general">BoxedObject</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Boolean.html?is-external=true" title="class or interface in java.lang">Boolean</a>&gt;&nbsp;smallMass)</code>
<div class="block">Returns a peptide from the peptide draft.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code><a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a></code></td>
<td class="colLast"><span class="typeNameLabel">ProteinIteratorUtils.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/protein_sequences/digestion/ProteinIteratorUtils.html#getPeptideFromProtein-char:A-int-double-double-">getPeptideFromProtein</a></span>(char[]&nbsp;proteinSequence,
                     int&nbsp;indexOnProtein,
                     double&nbsp;massMin,
                     double&nbsp;massMax)</code>
<div class="block">Returns a peptide from the given sequence.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code><a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a></code></td>
<td class="colLast"><span class="typeNameLabel">ProteinIteratorUtils.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/protein_sequences/digestion/ProteinIteratorUtils.html#getPeptideFromProtein-char:A-java.lang.String-int-java.lang.Double-java.lang.Double-">getPeptideFromProtein</a></span>(char[]&nbsp;peptideSequence,
                     <a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;proteinSequence,
                     int&nbsp;indexOnProtein,
                     <a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Double.html?is-external=true" title="class or interface in java.lang">Double</a>&nbsp;massMin,
                     <a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Double.html?is-external=true" title="class or interface in java.lang">Double</a>&nbsp;massMax)</code>
<div class="block">Returns a peptide from the given sequence on the given protein.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code><a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a></code></td>
<td class="colLast"><span class="typeNameLabel">ProteinIteratorUtils.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/protein_sequences/digestion/ProteinIteratorUtils.html#getPeptideFromProtein-char:A-java.lang.String-int-java.lang.Double-java.lang.Double-com.compomics.util.general.BoxedObject-">getPeptideFromProtein</a></span>(char[]&nbsp;peptideSequence,
                     <a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;proteinSequence,
                     int&nbsp;indexOnProtein,
                     <a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Double.html?is-external=true" title="class or interface in java.lang">Double</a>&nbsp;massMin,
                     <a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Double.html?is-external=true" title="class or interface in java.lang">Double</a>&nbsp;massMax,
                     <a href="../../../../../../../com/compomics/util/general/BoxedObject.html" title="class in com.compomics.util.general">BoxedObject</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Boolean.html?is-external=true" title="class or interface in java.lang">Boolean</a>&gt;&nbsp;smallMass)</code>
<div class="block">Returns a peptide from the given sequence on the given protein.</div>
</td>
</tr>
</tbody>
</table>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing constructors, and an explanation">
<caption><span>Constructors in <a href="../../../../../../../com/compomics/util/experiment/identification/protein_sequences/digestion/package-summary.html">com.compomics.util.experiment.identification.protein_sequences.digestion</a> with parameters of type <a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colOne" scope="col">Constructor and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/protein_sequences/digestion/PeptideWithPosition.html#PeptideWithPosition-com.compomics.util.experiment.biology.proteins.Peptide-int-">PeptideWithPosition</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;peptide,
                   int&nbsp;position)</code>
<div class="block">Constructor.</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="com.compomics.util.experiment.identification.psm_scoring">
<!--   -->
</a>
<h3>Uses of <a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a> in <a href="../../../../../../../com/compomics/util/experiment/identification/psm_scoring/package-summary.html">com.compomics.util.experiment.identification.psm_scoring</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing methods, and an explanation">
<caption><span>Methods in <a href="../../../../../../../com/compomics/util/experiment/identification/psm_scoring/package-summary.html">com.compomics.util.experiment.identification.psm_scoring</a> with parameters of type <a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>double</code></td>
<td class="colLast"><span class="typeNameLabel">PsmScoresEstimator.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/psm_scoring/PsmScoresEstimator.html#getDecreasingScore-com.compomics.util.experiment.biology.proteins.Peptide-java.lang.Integer-com.compomics.util.experiment.mass_spectrometry.spectra.Spectrum-com.compomics.util.parameters.identification.IdentificationParameters-com.compomics.util.experiment.identification.spectrum_annotation.SpecificAnnotationParameters-com.compomics.util.experiment.identification.spectrum_annotation.spectrum_annotators.PeptideSpectrumAnnotator-int-">getDecreasingScore</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;peptide,
                  <a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Integer.html?is-external=true" title="class or interface in java.lang">Integer</a>&nbsp;peptideCharge,
                  <a href="../../../../../../../com/compomics/util/experiment/mass_spectrometry/spectra/Spectrum.html" title="class in com.compomics.util.experiment.mass_spectrometry.spectra">Spectrum</a>&nbsp;spectrum,
                  <a href="../../../../../../../com/compomics/util/parameters/identification/IdentificationParameters.html" title="class in com.compomics.util.parameters.identification">IdentificationParameters</a>&nbsp;identificationParameters,
                  <a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/SpecificAnnotationParameters.html" title="class in com.compomics.util.experiment.identification.spectrum_annotation">SpecificAnnotationParameters</a>&nbsp;specificAnnotationPreferences,
                  <a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/spectrum_annotators/PeptideSpectrumAnnotator.html" title="class in com.compomics.util.experiment.identification.spectrum_annotation.spectrum_annotators">PeptideSpectrumAnnotator</a>&nbsp;peptideSpectrumAnnotator,
                  int&nbsp;scoreIndex)</code>
<div class="block">Scores the match between the given peptide and spectrum using the given
 score.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>double</code></td>
<td class="colLast"><span class="typeNameLabel">PsmScoresEstimator.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/psm_scoring/PsmScoresEstimator.html#getScore-com.compomics.util.experiment.biology.proteins.Peptide-java.lang.Integer-com.compomics.util.experiment.mass_spectrometry.spectra.Spectrum-com.compomics.util.parameters.identification.IdentificationParameters-com.compomics.util.experiment.identification.spectrum_annotation.SpecificAnnotationParameters-com.compomics.util.experiment.identification.spectrum_annotation.spectrum_annotators.PeptideSpectrumAnnotator-int-">getScore</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;peptide,
        <a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Integer.html?is-external=true" title="class or interface in java.lang">Integer</a>&nbsp;peptideCharge,
        <a href="../../../../../../../com/compomics/util/experiment/mass_spectrometry/spectra/Spectrum.html" title="class in com.compomics.util.experiment.mass_spectrometry.spectra">Spectrum</a>&nbsp;spectrum,
        <a href="../../../../../../../com/compomics/util/parameters/identification/IdentificationParameters.html" title="class in com.compomics.util.parameters.identification">IdentificationParameters</a>&nbsp;identificationParameters,
        <a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/SpecificAnnotationParameters.html" title="class in com.compomics.util.experiment.identification.spectrum_annotation">SpecificAnnotationParameters</a>&nbsp;specificAnnotationPreferences,
        <a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/spectrum_annotators/PeptideSpectrumAnnotator.html" title="class in com.compomics.util.experiment.identification.spectrum_annotation.spectrum_annotators">PeptideSpectrumAnnotator</a>&nbsp;peptideSpectrumAnnotator,
        int&nbsp;scoreIndex)</code>
<div class="block">Scores the match between the given peptide and spectrum using the given
 score.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>double</code></td>
<td class="colLast"><span class="typeNameLabel">PsmScoresEstimator.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/psm_scoring/PsmScoresEstimator.html#getScore-com.compomics.util.experiment.biology.proteins.Peptide-java.lang.Integer-com.compomics.util.experiment.mass_spectrometry.spectra.Spectrum-com.compomics.util.parameters.identification.IdentificationParameters-com.compomics.util.experiment.identification.spectrum_annotation.SpecificAnnotationParameters-com.compomics.util.experiment.identification.spectrum_annotation.spectrum_annotators.PeptideSpectrumAnnotator-com.compomics.util.experiment.identification.psm_scoring.PsmScore-">getScore</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;peptide,
        <a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Integer.html?is-external=true" title="class or interface in java.lang">Integer</a>&nbsp;peptideCharge,
        <a href="../../../../../../../com/compomics/util/experiment/mass_spectrometry/spectra/Spectrum.html" title="class in com.compomics.util.experiment.mass_spectrometry.spectra">Spectrum</a>&nbsp;spectrum,
        <a href="../../../../../../../com/compomics/util/parameters/identification/IdentificationParameters.html" title="class in com.compomics.util.parameters.identification">IdentificationParameters</a>&nbsp;identificationParameters,
        <a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/SpecificAnnotationParameters.html" title="class in com.compomics.util.experiment.identification.spectrum_annotation">SpecificAnnotationParameters</a>&nbsp;specificAnnotationPreferences,
        <a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/spectrum_annotators/PeptideSpectrumAnnotator.html" title="class in com.compomics.util.experiment.identification.spectrum_annotation.spectrum_annotators">PeptideSpectrumAnnotator</a>&nbsp;peptideSpectrumAnnotator,
        <a href="../../../../../../../com/compomics/util/experiment/identification/psm_scoring/PsmScore.html" title="enum in com.compomics.util.experiment.identification.psm_scoring">PsmScore</a>&nbsp;psmScore)</code>
<div class="block">Scores the match between the given peptide and spectrum using the given
 score.</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="com.compomics.util.experiment.identification.psm_scoring.psm_scores">
<!--   -->
</a>
<h3>Uses of <a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a> in <a href="../../../../../../../com/compomics/util/experiment/identification/psm_scoring/psm_scores/package-summary.html">com.compomics.util.experiment.identification.psm_scoring.psm_scores</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing methods, and an explanation">
<caption><span>Methods in <a href="../../../../../../../com/compomics/util/experiment/identification/psm_scoring/psm_scores/package-summary.html">com.compomics.util.experiment.identification.psm_scoring.psm_scores</a> with parameters of type <a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>double</code></td>
<td class="colLast"><span class="typeNameLabel">PrecursorAccuracy.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/psm_scoring/psm_scores/PrecursorAccuracy.html#getScore-com.compomics.util.experiment.biology.proteins.Peptide-int-com.compomics.util.experiment.mass_spectrometry.spectra.Precursor-boolean-int-int-">getScore</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;peptide,
        int&nbsp;identificationCharge,
        <a href="../../../../../../../com/compomics/util/experiment/mass_spectrometry/spectra/Precursor.html" title="class in com.compomics.util.experiment.mass_spectrometry.spectra">Precursor</a>&nbsp;precursor,
        boolean&nbsp;ppm,
        int&nbsp;minIsotope,
        int&nbsp;maxIsotope)</code>
<div class="block">Scores the match between the given peptide and spectrum using the
 precursor m/z accuracy.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>double</code></td>
<td class="colLast"><span class="typeNameLabel">HyperScore.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/psm_scoring/psm_scores/HyperScore.html#getScore-com.compomics.util.experiment.biology.proteins.Peptide-int-com.compomics.util.experiment.mass_spectrometry.spectra.Spectrum-java.util.ArrayList-">getScore</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;peptide,
        int&nbsp;charge,
        <a href="../../../../../../../com/compomics/util/experiment/mass_spectrometry/spectra/Spectrum.html" title="class in com.compomics.util.experiment.mass_spectrometry.spectra">Spectrum</a>&nbsp;spectrum,
        <a href="http://docs.oracle.com/javase/7/docs/api/java/util/ArrayList.html?is-external=true" title="class or interface in java.util">ArrayList</a>&lt;<a href="../../../../../../../com/compomics/util/experiment/identification/matches/IonMatch.html" title="class in com.compomics.util.experiment.identification.matches">IonMatch</a>&gt;&nbsp;ionMatches)</code>
<div class="block">Returns the hyperscore.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>double</code></td>
<td class="colLast"><span class="typeNameLabel">SnrScore.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/psm_scoring/psm_scores/SnrScore.html#getScore-com.compomics.util.experiment.biology.proteins.Peptide-com.compomics.util.experiment.mass_spectrometry.spectra.Spectrum-com.compomics.util.experiment.identification.spectrum_annotation.AnnotationParameters-com.compomics.util.experiment.identification.spectrum_annotation.SpecificAnnotationParameters-com.compomics.util.experiment.identification.spectrum_annotation.spectrum_annotators.PeptideSpectrumAnnotator-">getScore</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;peptide,
        <a href="../../../../../../../com/compomics/util/experiment/mass_spectrometry/spectra/Spectrum.html" title="class in com.compomics.util.experiment.mass_spectrometry.spectra">Spectrum</a>&nbsp;spectrum,
        <a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/AnnotationParameters.html" title="class in com.compomics.util.experiment.identification.spectrum_annotation">AnnotationParameters</a>&nbsp;annotationSettings,
        <a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/SpecificAnnotationParameters.html" title="class in com.compomics.util.experiment.identification.spectrum_annotation">SpecificAnnotationParameters</a>&nbsp;specificAnnotationSettings,
        <a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/spectrum_annotators/PeptideSpectrumAnnotator.html" title="class in com.compomics.util.experiment.identification.spectrum_annotation.spectrum_annotators">PeptideSpectrumAnnotator</a>&nbsp;peptideSpectrumAnnotator)</code>
<div class="block">Returns the score.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>double</code></td>
<td class="colLast"><span class="typeNameLabel">HyperScore.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/psm_scoring/psm_scores/HyperScore.html#getScore-com.compomics.util.experiment.biology.proteins.Peptide-com.compomics.util.experiment.mass_spectrometry.spectra.Spectrum-com.compomics.util.experiment.identification.spectrum_annotation.AnnotationParameters-com.compomics.util.experiment.identification.spectrum_annotation.SpecificAnnotationParameters-com.compomics.util.experiment.identification.spectrum_annotation.spectrum_annotators.PeptideSpectrumAnnotator-">getScore</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;peptide,
        <a href="../../../../../../../com/compomics/util/experiment/mass_spectrometry/spectra/Spectrum.html" title="class in com.compomics.util.experiment.mass_spectrometry.spectra">Spectrum</a>&nbsp;spectrum,
        <a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/AnnotationParameters.html" title="class in com.compomics.util.experiment.identification.spectrum_annotation">AnnotationParameters</a>&nbsp;annotationSettings,
        <a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/SpecificAnnotationParameters.html" title="class in com.compomics.util.experiment.identification.spectrum_annotation">SpecificAnnotationParameters</a>&nbsp;specificAnnotationSettings,
        <a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/spectrum_annotators/PeptideSpectrumAnnotator.html" title="class in com.compomics.util.experiment.identification.spectrum_annotation.spectrum_annotators">PeptideSpectrumAnnotator</a>&nbsp;peptideSpectrumAnnotator)</code>
<div class="block">Returns the hyperscore.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>double</code></td>
<td class="colLast"><span class="typeNameLabel">SnrScore.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/psm_scoring/psm_scores/SnrScore.html#getScore-com.compomics.util.experiment.biology.proteins.Peptide-com.compomics.util.experiment.mass_spectrometry.spectra.Spectrum-java.util.ArrayList-">getScore</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;peptide,
        <a href="../../../../../../../com/compomics/util/experiment/mass_spectrometry/spectra/Spectrum.html" title="class in com.compomics.util.experiment.mass_spectrometry.spectra">Spectrum</a>&nbsp;spectrum,
        <a href="http://docs.oracle.com/javase/7/docs/api/java/util/ArrayList.html?is-external=true" title="class or interface in java.util">ArrayList</a>&lt;<a href="../../../../../../../com/compomics/util/experiment/identification/matches/IonMatch.html" title="class in com.compomics.util.experiment.identification.matches">IonMatch</a>&gt;&nbsp;ionMatchesList)</code>
<div class="block">Returns the score.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>double</code></td>
<td class="colLast"><span class="typeNameLabel">SnrScore.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/psm_scoring/psm_scores/SnrScore.html#getScore-com.compomics.util.experiment.biology.proteins.Peptide-com.compomics.util.experiment.mass_spectrometry.spectra.Spectrum-java.util.HashMap-">getScore</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;peptide,
        <a href="../../../../../../../com/compomics/util/experiment/mass_spectrometry/spectra/Spectrum.html" title="class in com.compomics.util.experiment.mass_spectrometry.spectra">Spectrum</a>&nbsp;spectrum,
        <a href="http://docs.oracle.com/javase/7/docs/api/java/util/HashMap.html?is-external=true" title="class or interface in java.util">HashMap</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Double.html?is-external=true" title="class or interface in java.lang">Double</a>,<a href="http://docs.oracle.com/javase/7/docs/api/java/util/ArrayList.html?is-external=true" title="class or interface in java.util">ArrayList</a>&lt;<a href="../../../../../../../com/compomics/util/experiment/identification/matches/IonMatch.html" title="class in com.compomics.util.experiment.identification.matches">IonMatch</a>&gt;&gt;&nbsp;ionMatches)</code>
<div class="block">Returns the score.</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="com.compomics.util.experiment.identification.spectrum_annotation.simple_annotators">
<!--   -->
</a>
<h3>Uses of <a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a> in <a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/simple_annotators/package-summary.html">com.compomics.util.experiment.identification.spectrum_annotation.simple_annotators</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing constructors, and an explanation">
<caption><span>Constructors in <a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/simple_annotators/package-summary.html">com.compomics.util.experiment.identification.spectrum_annotation.simple_annotators</a> with parameters of type <a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colOne" scope="col">Constructor and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/simple_annotators/FragmentAnnotator.html#FragmentAnnotator-com.compomics.util.experiment.biology.proteins.Peptide-com.compomics.util.experiment.identification.spectrum_annotation.spectrum_annotators.SimplePeptideAnnotator.IonSeries-">FragmentAnnotator</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;peptide,
                 <a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/spectrum_annotators/SimplePeptideAnnotator.IonSeries.html" title="enum in com.compomics.util.experiment.identification.spectrum_annotation.spectrum_annotators">SimplePeptideAnnotator.IonSeries</a>&nbsp;ionSeries)</code>
<div class="block">Constructor.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/simple_annotators/FragmentAnnotator.html#FragmentAnnotator-com.compomics.util.experiment.biology.proteins.Peptide-com.compomics.util.experiment.identification.spectrum_annotation.spectrum_annotators.SimplePeptideAnnotator.IonSeries-boolean-boolean-">FragmentAnnotator</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;peptide,
                 <a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/spectrum_annotators/SimplePeptideAnnotator.IonSeries.html" title="enum in com.compomics.util.experiment.identification.spectrum_annotation.spectrum_annotators">SimplePeptideAnnotator.IonSeries</a>&nbsp;ionSeries,
                 boolean&nbsp;forward,
                 boolean&nbsp;complementary)</code>
<div class="block">Constructor.</div>
</td>
</tr>
<tr class="altColor">
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/simple_annotators/FragmentAnnotatorNL.html#FragmentAnnotatorNL-com.compomics.util.experiment.biology.proteins.Peptide-com.compomics.util.experiment.identification.spectrum_annotation.spectrum_annotators.SimplePeptideAnnotator.IonSeries-boolean-">FragmentAnnotatorNL</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;peptide,
                   <a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/spectrum_annotators/SimplePeptideAnnotator.IonSeries.html" title="enum in com.compomics.util.experiment.identification.spectrum_annotation.spectrum_annotators">SimplePeptideAnnotator.IonSeries</a>&nbsp;ionSeries,
                   boolean&nbsp;sequenceDependent)</code>
<div class="block">Constructor.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/simple_annotators/FragmentAnnotatorNL.html#FragmentAnnotatorNL-com.compomics.util.experiment.biology.proteins.Peptide-com.compomics.util.experiment.identification.spectrum_annotation.spectrum_annotators.SimplePeptideAnnotator.IonSeries-boolean-boolean-boolean-">FragmentAnnotatorNL</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;peptide,
                   <a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/spectrum_annotators/SimplePeptideAnnotator.IonSeries.html" title="enum in com.compomics.util.experiment.identification.spectrum_annotation.spectrum_annotators">SimplePeptideAnnotator.IonSeries</a>&nbsp;ionSeries,
                   boolean&nbsp;sequenceDependent,
                   boolean&nbsp;forward,
                   boolean&nbsp;complementary)</code>
<div class="block">Constructor.</div>
</td>
</tr>
<tr class="altColor">
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/simple_annotators/PrecursorAnnotator.html#PrecursorAnnotator-com.compomics.util.experiment.biology.proteins.Peptide-">PrecursorAnnotator</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;peptide)</code>
<div class="block">Constructor.</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="com.compomics.util.experiment.identification.spectrum_annotation.spectrum_annotators">
<!--   -->
</a>
<h3>Uses of <a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a> in <a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/spectrum_annotators/package-summary.html">com.compomics.util.experiment.identification.spectrum_annotation.spectrum_annotators</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing methods, and an explanation">
<caption><span>Methods in <a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/spectrum_annotators/package-summary.html">com.compomics.util.experiment.identification.spectrum_annotation.spectrum_annotators</a> that return <a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code><a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a></code></td>
<td class="colLast"><span class="typeNameLabel">PeptideSpectrumAnnotator.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/spectrum_annotators/PeptideSpectrumAnnotator.html#getCurrentlyLoadedPeptide--">getCurrentlyLoadedPeptide</a></span>()</code>
<div class="block">Returns the currently inspected peptide.</div>
</td>
</tr>
</tbody>
</table>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing methods, and an explanation">
<caption><span>Methods in <a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/spectrum_annotators/package-summary.html">com.compomics.util.experiment.identification.spectrum_annotation.spectrum_annotators</a> with parameters of type <a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code><a href="http://docs.oracle.com/javase/7/docs/api/java/util/Map.html?is-external=true" title="class or interface in java.util">Map</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Integer.html?is-external=true" title="class or interface in java.lang">Integer</a>,<a href="http://docs.oracle.com/javase/7/docs/api/java/util/ArrayList.html?is-external=true" title="class or interface in java.util">ArrayList</a>&lt;<a href="../../../../../../../com/compomics/util/experiment/identification/matches/IonMatch.html" title="class in com.compomics.util.experiment.identification.matches">IonMatch</a>&gt;&gt;</code></td>
<td class="colLast"><span class="typeNameLabel">PeptideSpectrumAnnotator.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/spectrum_annotators/PeptideSpectrumAnnotator.html#getCoveredAminoAcids-com.compomics.util.experiment.identification.spectrum_annotation.AnnotationParameters-com.compomics.util.experiment.identification.spectrum_annotation.SpecificAnnotationParameters-com.compomics.util.experiment.mass_spectrometry.spectra.Spectrum-com.compomics.util.experiment.biology.proteins.Peptide-boolean-">getCoveredAminoAcids</a></span>(<a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/AnnotationParameters.html" title="class in com.compomics.util.experiment.identification.spectrum_annotation">AnnotationParameters</a>&nbsp;annotationSettings,
                    <a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/SpecificAnnotationParameters.html" title="class in com.compomics.util.experiment.identification.spectrum_annotation">SpecificAnnotationParameters</a>&nbsp;specificAnnotationSettings,
                    <a href="../../../../../../../com/compomics/util/experiment/mass_spectrometry/spectra/Spectrum.html" title="class in com.compomics.util.experiment.mass_spectrometry.spectra">Spectrum</a>&nbsp;spectrum,
                    <a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;peptide,
                    boolean&nbsp;useIntensityFilter)</code>
<div class="block">Returns the ion matches corresponding to fragment ions indexed by amino
 acid number in the sequence.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static <a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/NeutralLossesMap.html" title="class in com.compomics.util.experiment.identification.spectrum_annotation">NeutralLossesMap</a></code></td>
<td class="colLast"><span class="typeNameLabel">PeptideSpectrumAnnotator.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/spectrum_annotators/PeptideSpectrumAnnotator.html#getDefaultLosses-com.compomics.util.experiment.biology.proteins.Peptide-">getDefaultLosses</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;peptide)</code>
<div class="block">Returns the possible neutral losses expected by default for a given
 peptide.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code><a href="http://docs.oracle.com/javase/7/docs/api/java/util/HashMap.html?is-external=true" title="class or interface in java.util">HashMap</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Integer.html?is-external=true" title="class or interface in java.lang">Integer</a>,<a href="http://docs.oracle.com/javase/7/docs/api/java/util/ArrayList.html?is-external=true" title="class or interface in java.util">ArrayList</a>&lt;<a href="../../../../../../../com/compomics/util/experiment/biology/ions/Ion.html" title="class in com.compomics.util.experiment.biology.ions">Ion</a>&gt;&gt;</code></td>
<td class="colLast"><span class="typeNameLabel">PeptideSpectrumAnnotator.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/spectrum_annotators/PeptideSpectrumAnnotator.html#getExpectedIons-com.compomics.util.experiment.identification.spectrum_annotation.SpecificAnnotationParameters-com.compomics.util.experiment.biology.proteins.Peptide-">getExpectedIons</a></span>(<a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/SpecificAnnotationParameters.html" title="class in com.compomics.util.experiment.identification.spectrum_annotation">SpecificAnnotationParameters</a>&nbsp;specificAnnotationSettings,
               <a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;peptide)</code>
<div class="block">Returns the expected ions in a map indexed by the possible charges.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code><a href="http://docs.oracle.com/javase/7/docs/api/java/util/HashMap.html?is-external=true" title="class or interface in java.util">HashMap</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Integer.html?is-external=true" title="class or interface in java.lang">Integer</a>,<a href="http://docs.oracle.com/javase/7/docs/api/java/util/ArrayList.html?is-external=true" title="class or interface in java.util">ArrayList</a>&lt;<a href="../../../../../../../com/compomics/util/experiment/biology/ions/Ion.html" title="class in com.compomics.util.experiment.biology.ions">Ion</a>&gt;&gt;</code></td>
<td class="colLast"><span class="typeNameLabel">PeptideSpectrumAnnotator.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/spectrum_annotators/PeptideSpectrumAnnotator.html#getExpectedIons-com.compomics.util.experiment.identification.spectrum_annotation.SpecificAnnotationParameters-com.compomics.util.experiment.biology.proteins.Peptide-java.util.HashMap-">getExpectedIons</a></span>(<a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/SpecificAnnotationParameters.html" title="class in com.compomics.util.experiment.identification.spectrum_annotation">SpecificAnnotationParameters</a>&nbsp;specificAnnotationSettings,
               <a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;peptide,
               <a href="http://docs.oracle.com/javase/7/docs/api/java/util/HashMap.html?is-external=true" title="class or interface in java.util">HashMap</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Integer.html?is-external=true" title="class or interface in java.lang">Integer</a>,<a href="http://docs.oracle.com/javase/7/docs/api/java/util/HashMap.html?is-external=true" title="class or interface in java.util">HashMap</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Integer.html?is-external=true" title="class or interface in java.lang">Integer</a>,<a href="http://docs.oracle.com/javase/7/docs/api/java/util/ArrayList.html?is-external=true" title="class or interface in java.util">ArrayList</a>&lt;<a href="../../../../../../../com/compomics/util/experiment/biology/ions/Ion.html" title="class in com.compomics.util.experiment.biology.ions">Ion</a>&gt;&gt;&gt;&nbsp;possibleFragmentIons)</code>
<div class="block">Returns the expected ions in a map indexed by the possible charges.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code><a href="http://docs.oracle.com/javase/7/docs/api/java/util/stream.Stream.html?is-external=true" title="class or interface in java.util">java.util.stream.Stream</a>&lt;<a href="../../../../../../../com/compomics/util/experiment/identification/matches/IonMatch.html" title="class in com.compomics.util.experiment.identification.matches">IonMatch</a>&gt;</code></td>
<td class="colLast"><span class="typeNameLabel">PeptideSpectrumAnnotator.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/spectrum_annotators/PeptideSpectrumAnnotator.html#getSpectrumAnnotation-com.compomics.util.experiment.identification.spectrum_annotation.AnnotationParameters-com.compomics.util.experiment.identification.spectrum_annotation.SpecificAnnotationParameters-com.compomics.util.experiment.mass_spectrometry.spectra.Spectrum-com.compomics.util.experiment.biology.proteins.Peptide-">getSpectrumAnnotation</a></span>(<a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/AnnotationParameters.html" title="class in com.compomics.util.experiment.identification.spectrum_annotation">AnnotationParameters</a>&nbsp;annotationSettings,
                     <a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/SpecificAnnotationParameters.html" title="class in com.compomics.util.experiment.identification.spectrum_annotation">SpecificAnnotationParameters</a>&nbsp;specificAnnotationSettings,
                     <a href="../../../../../../../com/compomics/util/experiment/mass_spectrometry/spectra/Spectrum.html" title="class in com.compomics.util.experiment.mass_spectrometry.spectra">Spectrum</a>&nbsp;spectrum,
                     <a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;peptide)</code>
<div class="block">Returns the spectrum annotations of a spectrum in a list of IonMatches
 using the intensity filter.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code><a href="http://docs.oracle.com/javase/7/docs/api/java/util/ArrayList.html?is-external=true" title="class or interface in java.util">ArrayList</a>&lt;<a href="../../../../../../../com/compomics/util/experiment/identification/matches/IonMatch.html" title="class in com.compomics.util.experiment.identification.matches">IonMatch</a>&gt;</code></td>
<td class="colLast"><span class="typeNameLabel">PeptideSpectrumAnnotator.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/spectrum_annotators/PeptideSpectrumAnnotator.html#getSpectrumAnnotation-com.compomics.util.experiment.identification.spectrum_annotation.AnnotationParameters-com.compomics.util.experiment.identification.spectrum_annotation.SpecificAnnotationParameters-com.compomics.util.experiment.mass_spectrometry.spectra.Spectrum-com.compomics.util.experiment.biology.proteins.Peptide-boolean-">getSpectrumAnnotation</a></span>(<a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/AnnotationParameters.html" title="class in com.compomics.util.experiment.identification.spectrum_annotation">AnnotationParameters</a>&nbsp;annotationSettings,
                     <a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/SpecificAnnotationParameters.html" title="class in com.compomics.util.experiment.identification.spectrum_annotation">SpecificAnnotationParameters</a>&nbsp;specificAnnotationSettings,
                     <a href="../../../../../../../com/compomics/util/experiment/mass_spectrometry/spectra/Spectrum.html" title="class in com.compomics.util.experiment.mass_spectrometry.spectra">Spectrum</a>&nbsp;spectrum,
                     <a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;peptide,
                     boolean&nbsp;useIntensityFilter)</code>
<div class="block">Returns the spectrum annotations of a spectrum in a list of IonMatches.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code><a href="http://docs.oracle.com/javase/7/docs/api/java/util/stream.Stream.html?is-external=true" title="class or interface in java.util">java.util.stream.Stream</a>&lt;<a href="../../../../../../../com/compomics/util/experiment/identification/matches/IonMatch.html" title="class in com.compomics.util.experiment.identification.matches">IonMatch</a>&gt;</code></td>
<td class="colLast"><span class="typeNameLabel">PeptideSpectrumAnnotator.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/spectrum_annotators/PeptideSpectrumAnnotator.html#getSpectrumAnnotation-com.compomics.util.experiment.identification.spectrum_annotation.AnnotationParameters-com.compomics.util.experiment.identification.spectrum_annotation.SpecificAnnotationParameters-com.compomics.util.experiment.mass_spectrometry.spectra.Spectrum-com.compomics.util.experiment.biology.proteins.Peptide-java.util.HashMap-boolean-">getSpectrumAnnotation</a></span>(<a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/AnnotationParameters.html" title="class in com.compomics.util.experiment.identification.spectrum_annotation">AnnotationParameters</a>&nbsp;annotationSettings,
                     <a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/SpecificAnnotationParameters.html" title="class in com.compomics.util.experiment.identification.spectrum_annotation">SpecificAnnotationParameters</a>&nbsp;specificAnnotationSettings,
                     <a href="../../../../../../../com/compomics/util/experiment/mass_spectrometry/spectra/Spectrum.html" title="class in com.compomics.util.experiment.mass_spectrometry.spectra">Spectrum</a>&nbsp;spectrum,
                     <a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;peptide,
                     <a href="http://docs.oracle.com/javase/7/docs/api/java/util/HashMap.html?is-external=true" title="class or interface in java.util">HashMap</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Integer.html?is-external=true" title="class or interface in java.lang">Integer</a>,<a href="http://docs.oracle.com/javase/7/docs/api/java/util/HashMap.html?is-external=true" title="class or interface in java.util">HashMap</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Integer.html?is-external=true" title="class or interface in java.lang">Integer</a>,<a href="http://docs.oracle.com/javase/7/docs/api/java/util/ArrayList.html?is-external=true" title="class or interface in java.util">ArrayList</a>&lt;<a href="../../../../../../../com/compomics/util/experiment/biology/ions/Ion.html" title="class in com.compomics.util.experiment.biology.ions">Ion</a>&gt;&gt;&gt;&nbsp;possiblePeptideFragments,
                     boolean&nbsp;useIntensityFilter)</code>
<div class="block">Returns the spectrum annotations of a spectrum in a list of IonMatches.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code><a href="http://docs.oracle.com/javase/7/docs/api/java/util/stream.Stream.html?is-external=true" title="class or interface in java.util">java.util.stream.Stream</a>&lt;<a href="../../../../../../../com/compomics/util/experiment/identification/matches/IonMatch.html" title="class in com.compomics.util.experiment.identification.matches">IonMatch</a>&gt;</code></td>
<td class="colLast"><span class="typeNameLabel">PeptideSpectrumAnnotator.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/spectrum_annotators/PeptideSpectrumAnnotator.html#getSpectrumAnnotationStream-com.compomics.util.experiment.identification.spectrum_annotation.AnnotationParameters-com.compomics.util.experiment.identification.spectrum_annotation.SpecificAnnotationParameters-com.compomics.util.experiment.mass_spectrometry.spectra.Spectrum-com.compomics.util.experiment.biology.proteins.Peptide-boolean-">getSpectrumAnnotationStream</a></span>(<a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/AnnotationParameters.html" title="class in com.compomics.util.experiment.identification.spectrum_annotation">AnnotationParameters</a>&nbsp;annotationSettings,
                           <a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/SpecificAnnotationParameters.html" title="class in com.compomics.util.experiment.identification.spectrum_annotation">SpecificAnnotationParameters</a>&nbsp;specificAnnotationSettings,
                           <a href="../../../../../../../com/compomics/util/experiment/mass_spectrometry/spectra/Spectrum.html" title="class in com.compomics.util.experiment.mass_spectrometry.spectra">Spectrum</a>&nbsp;spectrum,
                           <a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;peptide,
                           boolean&nbsp;useIntensityFilter)</code>
<div class="block">Returns the spectrum annotations of a spectrum in a list of IonMatches.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code><a href="http://docs.oracle.com/javase/7/docs/api/java/util/ArrayList.html?is-external=true" title="class or interface in java.util">ArrayList</a>&lt;<a href="../../../../../../../com/compomics/util/experiment/identification/matches/IonMatch.html" title="class in com.compomics.util.experiment.identification.matches">IonMatch</a>&gt;</code></td>
<td class="colLast"><span class="typeNameLabel">PeptideSpectrumAnnotator.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/spectrum_annotators/PeptideSpectrumAnnotator.html#matchPeak-com.compomics.util.experiment.biology.proteins.Peptide-com.compomics.util.experiment.identification.spectrum_annotation.SpecificAnnotationParameters-com.compomics.util.experiment.mass_spectrometry.spectra.Peak-">matchPeak</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;peptide,
         <a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/SpecificAnnotationParameters.html" title="class in com.compomics.util.experiment.identification.spectrum_annotation">SpecificAnnotationParameters</a>&nbsp;specificAnnotationSettings,
         <a href="../../../../../../../com/compomics/util/experiment/mass_spectrometry/spectra/Peak.html" title="class in com.compomics.util.experiment.mass_spectrometry.spectra">Peak</a>&nbsp;peak)</code>
<div class="block">This method matches the potential fragment ions of a given peptide with a
 given peak according to the annotation settings.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><span class="typeNameLabel">PeptideSpectrumAnnotator.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/spectrum_annotators/PeptideSpectrumAnnotator.html#setPeptide-com.compomics.util.experiment.biology.proteins.Peptide-java.util.HashMap-int-com.compomics.util.experiment.identification.spectrum_annotation.SpecificAnnotationParameters-">setPeptide</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;peptide,
          <a href="http://docs.oracle.com/javase/7/docs/api/java/util/HashMap.html?is-external=true" title="class or interface in java.util">HashMap</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Integer.html?is-external=true" title="class or interface in java.lang">Integer</a>,<a href="http://docs.oracle.com/javase/7/docs/api/java/util/HashMap.html?is-external=true" title="class or interface in java.util">HashMap</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Integer.html?is-external=true" title="class or interface in java.lang">Integer</a>,<a href="http://docs.oracle.com/javase/7/docs/api/java/util/ArrayList.html?is-external=true" title="class or interface in java.util">ArrayList</a>&lt;<a href="../../../../../../../com/compomics/util/experiment/biology/ions/Ion.html" title="class in com.compomics.util.experiment.biology.ions">Ion</a>&gt;&gt;&gt;&nbsp;possibleFragmentIons,
          int&nbsp;precursorCharge,
          <a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/SpecificAnnotationParameters.html" title="class in com.compomics.util.experiment.identification.spectrum_annotation">SpecificAnnotationParameters</a>&nbsp;specificAnnotationSettings)</code>
<div class="block">Sets a new peptide to annotate.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><span class="typeNameLabel">PeptideSpectrumAnnotator.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/spectrum_annotators/PeptideSpectrumAnnotator.html#setPeptide-com.compomics.util.experiment.biology.proteins.Peptide-int-com.compomics.util.experiment.identification.spectrum_annotation.SpecificAnnotationParameters-">setPeptide</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;peptide,
          int&nbsp;precursorCharge,
          <a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/SpecificAnnotationParameters.html" title="class in com.compomics.util.experiment.identification.spectrum_annotation">SpecificAnnotationParameters</a>&nbsp;specificAnnotationSettings)</code>
<div class="block">Sets a new peptide to annotate.</div>
</td>
</tr>
</tbody>
</table>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing constructors, and an explanation">
<caption><span>Constructors in <a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/spectrum_annotators/package-summary.html">com.compomics.util.experiment.identification.spectrum_annotation.spectrum_annotators</a> with parameters of type <a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colOne" scope="col">Constructor and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/spectrum_annotators/SimplePeptideAnnotator.html#SimplePeptideAnnotator-com.compomics.util.experiment.biology.proteins.Peptide-int-com.compomics.util.experiment.identification.spectrum_annotation.AnnotationParameters-">SimplePeptideAnnotator</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;peptide,
                      int&nbsp;charge,
                      <a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/AnnotationParameters.html" title="class in com.compomics.util.experiment.identification.spectrum_annotation">AnnotationParameters</a>&nbsp;annotationSettings)</code>
<div class="block">Constructor.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/spectrum_annotators/SimplePeptideAnnotator.html#SimplePeptideAnnotator-com.compomics.util.experiment.biology.proteins.Peptide-int-boolean-boolean-boolean-boolean-boolean-boolean-boolean-boolean-boolean-boolean-boolean-boolean-com.compomics.util.experiment.biology.ions.impl.ReporterIon:A-">SimplePeptideAnnotator</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;peptide,
                      int&nbsp;charge,
                      boolean&nbsp;a,
                      boolean&nbsp;b,
                      boolean&nbsp;c,
                      boolean&nbsp;x,
                      boolean&nbsp;y,
                      boolean&nbsp;z,
                      boolean&nbsp;precursor,
                      boolean&nbsp;immonium,
                      boolean&nbsp;related,
                      boolean&nbsp;reporter,
                      boolean&nbsp;neutralLosses,
                      boolean&nbsp;neutralLossesSequenceDependent,
                      <a href="../../../../../../../com/compomics/util/experiment/biology/ions/impl/ReporterIon.html" title="class in com.compomics.util.experiment.biology.ions.impl">ReporterIon</a>[]&nbsp;reporterIons)</code>
<div class="block">Constructor.</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="com.compomics.util.experiment.identification.spectrum_assumptions">
<!--   -->
</a>
<h3>Uses of <a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a> in <a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_assumptions/package-summary.html">com.compomics.util.experiment.identification.spectrum_assumptions</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing methods, and an explanation">
<caption><span>Methods in <a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_assumptions/package-summary.html">com.compomics.util.experiment.identification.spectrum_assumptions</a> that return <a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code><a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a></code></td>
<td class="colLast"><span class="typeNameLabel">PeptideAssumption.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_assumptions/PeptideAssumption.html#getPeptide--">getPeptide</a></span>()</code>
<div class="block">Get the theoretic peptide.</div>
</td>
</tr>
</tbody>
</table>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing methods, and an explanation">
<caption><span>Methods in <a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_assumptions/package-summary.html">com.compomics.util.experiment.identification.spectrum_assumptions</a> with parameters of type <a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><span class="typeNameLabel">PeptideAssumption.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_assumptions/PeptideAssumption.html#setPeptide-com.compomics.util.experiment.biology.proteins.Peptide-">setPeptide</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;peptide)</code>&nbsp;</td>
</tr>
</tbody>
</table>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing constructors, and an explanation">
<caption><span>Constructors in <a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_assumptions/package-summary.html">com.compomics.util.experiment.identification.spectrum_assumptions</a> with parameters of type <a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colOne" scope="col">Constructor and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_assumptions/PeptideAssumption.html#PeptideAssumption-com.compomics.util.experiment.biology.proteins.Peptide-int-">PeptideAssumption</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;peptide,
                 int&nbsp;identificationCharge)</code>
<div class="block">Constructor for a simple peptide assumption containing only the
 information necessary for spectrum annotation.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_assumptions/PeptideAssumption.html#PeptideAssumption-com.compomics.util.experiment.biology.proteins.Peptide-int-int-int-double-">PeptideAssumption</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;peptide,
                 int&nbsp;rank,
                 int&nbsp;advocate,
                 int&nbsp;identificationCharge,
                 double&nbsp;score)</code>
<div class="block">Constructor for a peptide assumption.</div>
</td>
</tr>
<tr class="altColor">
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_assumptions/PeptideAssumption.html#PeptideAssumption-com.compomics.util.experiment.biology.proteins.Peptide-int-int-int-double-java.lang.String-">PeptideAssumption</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;peptide,
                 int&nbsp;rank,
                 int&nbsp;advocate,
                 int&nbsp;identificationCharge,
                 double&nbsp;score,
                 <a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;identificationFile)</code>
<div class="block">Constructor for a peptide assumption.</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="com.compomics.util.experiment.identification.utils">
<!--   -->
</a>
<h3>Uses of <a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a> in <a href="../../../../../../../com/compomics/util/experiment/identification/utils/package-summary.html">com.compomics.util.experiment.identification.utils</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing methods, and an explanation">
<caption><span>Methods in <a href="../../../../../../../com/compomics/util/experiment/identification/utils/package-summary.html">com.compomics.util.experiment.identification.utils</a> with parameters of type <a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>static <a href="http://docs.oracle.com/javase/7/docs/api/java/util/TreeMap.html?is-external=true" title="class or interface in java.util">TreeMap</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>,<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>[]&gt;</code></td>
<td class="colLast"><span class="typeNameLabel">PeptideUtils.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/utils/PeptideUtils.html#getAaAfter-com.compomics.util.experiment.biology.proteins.Peptide-int-com.compomics.util.experiment.io.biology.protein.SequenceProvider-">getAaAfter</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;peptide,
          int&nbsp;nAa,
          <a href="../../../../../../../com/compomics/util/experiment/io/biology/protein/SequenceProvider.html" title="interface in com.compomics.util.experiment.io.biology.protein">SequenceProvider</a>&nbsp;sequenceProvider)</code>
<div class="block">Returns the amino acids before the given peptide as a string in a map
 based on the peptide protein mapping.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static <a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a></code></td>
<td class="colLast"><span class="typeNameLabel">PeptideUtils.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/utils/PeptideUtils.html#getAaAfter-com.compomics.util.experiment.biology.proteins.Peptide-java.lang.String-int-int-com.compomics.util.experiment.io.biology.protein.SequenceProvider-">getAaAfter</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;peptide,
          <a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;accession,
          int&nbsp;index,
          int&nbsp;nAa,
          <a href="../../../../../../../com/compomics/util/experiment/io/biology/protein/SequenceProvider.html" title="interface in com.compomics.util.experiment.io.biology.protein">SequenceProvider</a>&nbsp;sequenceProvider)</code>
<div class="block">Returns the amino acids before the given peptide as a string.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static <a href="http://docs.oracle.com/javase/7/docs/api/java/util/TreeMap.html?is-external=true" title="class or interface in java.util">TreeMap</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>,<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>[]&gt;</code></td>
<td class="colLast"><span class="typeNameLabel">PeptideUtils.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/utils/PeptideUtils.html#getAaBefore-com.compomics.util.experiment.biology.proteins.Peptide-int-com.compomics.util.experiment.io.biology.protein.SequenceProvider-">getAaBefore</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;peptide,
           int&nbsp;nAa,
           <a href="../../../../../../../com/compomics/util/experiment/io/biology/protein/SequenceProvider.html" title="interface in com.compomics.util.experiment.io.biology.protein">SequenceProvider</a>&nbsp;sequenceProvider)</code>
<div class="block">Returns the amino acids before the given peptide as a string in a map
 based on the peptide protein mapping.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static <a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a></code></td>
<td class="colLast"><span class="typeNameLabel">PeptideUtils.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/utils/PeptideUtils.html#getAaBefore-com.compomics.util.experiment.biology.proteins.Peptide-java.lang.String-int-int-com.compomics.util.experiment.io.biology.protein.SequenceProvider-">getAaBefore</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;peptide,
           <a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;accession,
           int&nbsp;index,
           int&nbsp;nAa,
           <a href="../../../../../../../com/compomics/util/experiment/io/biology/protein/SequenceProvider.html" title="interface in com.compomics.util.experiment.io.biology.protein">SequenceProvider</a>&nbsp;sequenceProvider)</code>
<div class="block">Returns the amino acids before the given peptide as a string in a map
 based on the peptide protein mapping.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static <a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a></code></td>
<td class="colLast"><span class="typeNameLabel">PeptideUtils.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/utils/PeptideUtils.html#getPeptideModificationsAsString-com.compomics.util.experiment.biology.proteins.Peptide-boolean-">getPeptideModificationsAsString</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;peptide,
                               boolean&nbsp;variable)</code>
<div class="block">Returns the peptide modifications as a string.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static <a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a></code></td>
<td class="colLast"><span class="typeNameLabel">PeptideUtils.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/utils/PeptideUtils.html#getTaggedModifiedSequence-com.compomics.util.experiment.biology.proteins.Peptide-com.compomics.util.parameters.identification.search.ModificationParameters-java.util.HashMap-java.util.HashMap-java.util.HashMap-java.util.HashMap-boolean-boolean-boolean-">getTaggedModifiedSequence</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;peptide,
                         <a href="../../../../../../../com/compomics/util/parameters/identification/search/ModificationParameters.html" title="class in com.compomics.util.parameters.identification.search">ModificationParameters</a>&nbsp;modificationProfile,
                         <a href="http://docs.oracle.com/javase/7/docs/api/java/util/HashMap.html?is-external=true" title="class or interface in java.util">HashMap</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Integer.html?is-external=true" title="class or interface in java.lang">Integer</a>,<a href="http://docs.oracle.com/javase/7/docs/api/java/util/ArrayList.html?is-external=true" title="class or interface in java.util">ArrayList</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&gt;&gt;&nbsp;confidentModificationSites,
                         <a href="http://docs.oracle.com/javase/7/docs/api/java/util/HashMap.html?is-external=true" title="class or interface in java.util">HashMap</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Integer.html?is-external=true" title="class or interface in java.lang">Integer</a>,<a href="http://docs.oracle.com/javase/7/docs/api/java/util/ArrayList.html?is-external=true" title="class or interface in java.util">ArrayList</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&gt;&gt;&nbsp;representativeAmbiguousModificationSites,
                         <a href="http://docs.oracle.com/javase/7/docs/api/java/util/HashMap.html?is-external=true" title="class or interface in java.util">HashMap</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Integer.html?is-external=true" title="class or interface in java.lang">Integer</a>,<a href="http://docs.oracle.com/javase/7/docs/api/java/util/ArrayList.html?is-external=true" title="class or interface in java.util">ArrayList</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&gt;&gt;&nbsp;secondaryAmbiguousModificationSites,
                         <a href="http://docs.oracle.com/javase/7/docs/api/java/util/HashMap.html?is-external=true" title="class or interface in java.util">HashMap</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Integer.html?is-external=true" title="class or interface in java.lang">Integer</a>,<a href="http://docs.oracle.com/javase/7/docs/api/java/util/ArrayList.html?is-external=true" title="class or interface in java.util">ArrayList</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&gt;&gt;&nbsp;fixedModificationSites,
                         boolean&nbsp;useHtmlColorCoding,
                         boolean&nbsp;includeHtmlStartEndTags,
                         boolean&nbsp;useShortName)</code>
<div class="block">Returns the modified sequence as an tagged string with potential
 modification sites color coded or with Modification tags, e.g,
 &lt;mox&gt;.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static boolean</code></td>
<td class="colLast"><span class="typeNameLabel">PeptideUtils.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/utils/PeptideUtils.html#isDecoy-com.compomics.util.experiment.biology.proteins.Peptide-com.compomics.util.experiment.io.biology.protein.SequenceProvider-">isDecoy</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;peptide,
       <a href="../../../../../../../com/compomics/util/experiment/io/biology/protein/SequenceProvider.html" title="interface in com.compomics.util.experiment.io.biology.protein">SequenceProvider</a>&nbsp;sequenceProvider)</code>
<div class="block">Returns a boolean indicating whether the peptide matches a decoy
 sequence.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static boolean</code></td>
<td class="colLast"><span class="typeNameLabel">PeptideUtils.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/utils/PeptideUtils.html#isEnzymatic-com.compomics.util.experiment.biology.proteins.Peptide-com.compomics.util.experiment.io.biology.protein.SequenceProvider-java.util.ArrayList-">isEnzymatic</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;peptide,
           <a href="../../../../../../../com/compomics/util/experiment/io/biology/protein/SequenceProvider.html" title="interface in com.compomics.util.experiment.io.biology.protein">SequenceProvider</a>&nbsp;sequenceProvider,
           <a href="http://docs.oracle.com/javase/7/docs/api/java/util/ArrayList.html?is-external=true" title="class or interface in java.util">ArrayList</a>&lt;<a href="../../../../../../../com/compomics/util/experiment/biology/enzymes/Enzyme.html" title="class in com.compomics.util.experiment.biology.enzymes">Enzyme</a>&gt;&nbsp;enzymes)</code>
<div class="block">Returns a boolean indicating whether the peptide is enzymatic in at least one protein using one of the given enzymes.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static boolean</code></td>
<td class="colLast"><span class="typeNameLabel">PeptideUtils.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/utils/PeptideUtils.html#isEnzymatic-com.compomics.util.experiment.biology.proteins.Peptide-java.lang.String-java.lang.String-java.util.ArrayList-">isEnzymatic</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;peptide,
           <a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;proteinAccession,
           <a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;proteinSequence,
           <a href="http://docs.oracle.com/javase/7/docs/api/java/util/ArrayList.html?is-external=true" title="class or interface in java.util">ArrayList</a>&lt;<a href="../../../../../../../com/compomics/util/experiment/biology/enzymes/Enzyme.html" title="class in com.compomics.util.experiment.biology.enzymes">Enzyme</a>&gt;&nbsp;enzymes)</code>
<div class="block">Returns a boolean indicating whether the peptide is enzymatic using one of the given enzymes.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static boolean</code></td>
<td class="colLast"><span class="typeNameLabel">PeptideUtils.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/experiment/identification/utils/PeptideUtils.html#isVariant-com.compomics.util.experiment.biology.proteins.Peptide-java.lang.String-">isVariant</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;peptide,
         <a href="http://docs.oracle.com/javase/7/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;accession)</code>
<div class="block">Returns a boolean indicating whether the peptide needs variants to be mapped to the given protein.</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="com.compomics.util.gui.spectrum">
<!--   -->
</a>
<h3>Uses of <a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a> in <a href="../../../../../../../com/compomics/util/gui/spectrum/package-summary.html">com.compomics.util.gui.spectrum</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing methods, and an explanation">
<caption><span>Methods in <a href="../../../../../../../com/compomics/util/gui/spectrum/package-summary.html">com.compomics.util.gui.spectrum</a> with parameters of type <a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><span class="typeNameLabel">SpectrumPanel.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/gui/spectrum/SpectrumPanel.html#addAutomaticDeNovoSequencing-com.compomics.util.experiment.biology.proteins.Peptide-java.util.ArrayList-int-int-int-boolean-boolean-boolean-">addAutomaticDeNovoSequencing</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;currentPeptide,
                            <a href="http://docs.oracle.com/javase/7/docs/api/java/util/ArrayList.html?is-external=true" title="class or interface in java.util">ArrayList</a>&lt;<a href="../../../../../../../com/compomics/util/experiment/identification/matches/IonMatch.html" title="class in com.compomics.util.experiment.identification.matches">IonMatch</a>&gt;&nbsp;annotations,
                            int&nbsp;aForwardIon,
                            int&nbsp;aRewindIon,
                            int&nbsp;aDeNovoCharge,
                            boolean&nbsp;showForwardTags,
                            boolean&nbsp;showRewindTags,
                            boolean&nbsp;mirrored)</code>
<div class="block">Add reference areas annotating the de novo tags, using default percent
 height of 0.9 for the forward ions and 1.0 for the reverse ions default
 alpha levels of 0.2.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><span class="typeNameLabel">SpectrumPanel.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/gui/spectrum/SpectrumPanel.html#addAutomaticDeNovoSequencing-com.compomics.util.experiment.biology.proteins.Peptide-java.util.ArrayList-int-int-int-boolean-boolean-double-double-java.util.ArrayList-boolean-boolean-">addAutomaticDeNovoSequencing</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;currentPeptide,
                            <a href="http://docs.oracle.com/javase/7/docs/api/java/util/ArrayList.html?is-external=true" title="class or interface in java.util">ArrayList</a>&lt;<a href="../../../../../../../com/compomics/util/experiment/identification/matches/IonMatch.html" title="class in com.compomics.util.experiment.identification.matches">IonMatch</a>&gt;&nbsp;annotations,
                            int&nbsp;aForwardIon,
                            int&nbsp;aRewindIon,
                            int&nbsp;aDeNovoCharge,
                            boolean&nbsp;showForwardTags,
                            boolean&nbsp;showRewindTags,
                            double&nbsp;forwardIonPercentHeight,
                            double&nbsp;rewindIonPercentHeight,
                            <a href="http://docs.oracle.com/javase/7/docs/api/java/util/ArrayList.html?is-external=true" title="class or interface in java.util">ArrayList</a>&lt;float[]&gt;&nbsp;alphaLevels,
                            boolean&nbsp;excludeFixedModifications,
                            boolean&nbsp;mirrored)</code>
<div class="block">Add reference areas annotating the de novo tags.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><span class="typeNameLabel">SpectrumPanel.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/gui/spectrum/SpectrumPanel.html#addAutomaticDeNovoSequencing-com.compomics.util.experiment.biology.proteins.Peptide-java.util.ArrayList-int-int-int-boolean-boolean-double-double-boolean-">addAutomaticDeNovoSequencing</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;currentPeptide,
                            <a href="http://docs.oracle.com/javase/7/docs/api/java/util/ArrayList.html?is-external=true" title="class or interface in java.util">ArrayList</a>&lt;<a href="../../../../../../../com/compomics/util/experiment/identification/matches/IonMatch.html" title="class in com.compomics.util.experiment.identification.matches">IonMatch</a>&gt;&nbsp;annotations,
                            int&nbsp;aForwardIon,
                            int&nbsp;aRewindIon,
                            int&nbsp;aDeNovoCharge,
                            boolean&nbsp;showForwardTags,
                            boolean&nbsp;showRewindTags,
                            double&nbsp;forwardIonPercentHeight,
                            double&nbsp;rewindIonPercentHeight,
                            boolean&nbsp;mirrored)</code>
<div class="block">Add reference areas annotating the de novo tags, using default alpha
 levels of 0.2.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><span class="typeNameLabel">SpectrumPanel.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/gui/spectrum/SpectrumPanel.html#addAutomaticDeNovoSequencing-com.compomics.util.experiment.biology.proteins.Peptide-java.util.ArrayList-int-int-int-boolean-boolean-double-double-boolean-boolean-">addAutomaticDeNovoSequencing</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;currentPeptide,
                            <a href="http://docs.oracle.com/javase/7/docs/api/java/util/ArrayList.html?is-external=true" title="class or interface in java.util">ArrayList</a>&lt;<a href="../../../../../../../com/compomics/util/experiment/identification/matches/IonMatch.html" title="class in com.compomics.util.experiment.identification.matches">IonMatch</a>&gt;&nbsp;annotations,
                            int&nbsp;aForwardIon,
                            int&nbsp;aRewindIon,
                            int&nbsp;aDeNovoCharge,
                            boolean&nbsp;showForwardTags,
                            boolean&nbsp;showRewindTags,
                            double&nbsp;forwardIonPercentHeight,
                            double&nbsp;rewindIonPercentHeight,
                            boolean&nbsp;excludeFixedModifications,
                            boolean&nbsp;mirrored)</code>
<div class="block">Add reference areas annotating the de novo tags, using default alpha
 levels of 0.2.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><span class="typeNameLabel">SpectrumPanel.</span><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/gui/spectrum/SpectrumPanel.html#addAutomaticDeNovoSequencing-com.compomics.util.experiment.biology.proteins.Peptide-java.util.ArrayList-int-int-int-boolean-boolean-double-double-float-float-java.util.ArrayList-boolean-boolean-">addAutomaticDeNovoSequencing</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;currentPeptide,
                            <a href="http://docs.oracle.com/javase/7/docs/api/java/util/ArrayList.html?is-external=true" title="class or interface in java.util">ArrayList</a>&lt;<a href="../../../../../../../com/compomics/util/experiment/identification/matches/IonMatch.html" title="class in com.compomics.util.experiment.identification.matches">IonMatch</a>&gt;&nbsp;annotations,
                            int&nbsp;aForwardIon,
                            int&nbsp;aRewindIon,
                            int&nbsp;aDeNovoCharge,
                            boolean&nbsp;showForwardTags,
                            boolean&nbsp;showRewindTags,
                            double&nbsp;forwardIonPercentHeight,
                            double&nbsp;rewindIonPercentHeight,
                            float&nbsp;forwardIonAlphaLevel,
                            float&nbsp;rewindIonAlphaLevel,
                            <a href="http://docs.oracle.com/javase/7/docs/api/java/util/ArrayList.html?is-external=true" title="class or interface in java.util">ArrayList</a>&lt;float[]&gt;&nbsp;alphaLevels,
                            boolean&nbsp;excludeFixedModifications,
                            boolean&nbsp;mirrored)</code>
<div class="block">Add reference areas annotating the de novo tags.</div>
</td>
</tr>
</tbody>
</table>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing constructors, and an explanation">
<caption><span>Constructors in <a href="../../../../../../../com/compomics/util/gui/spectrum/package-summary.html">com.compomics.util.gui.spectrum</a> with parameters of type <a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colOne" scope="col">Constructor and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/gui/spectrum/FragmentIonTable.html#FragmentIonTable-com.compomics.util.experiment.biology.proteins.Peptide-java.util.ArrayList-java.util.ArrayList-java.util.HashSet-com.compomics.util.experiment.identification.spectrum_annotation.NeutralLossesMap-boolean-boolean-">FragmentIonTable</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;currentPeptide,
                <a href="http://docs.oracle.com/javase/7/docs/api/java/util/ArrayList.html?is-external=true" title="class or interface in java.util">ArrayList</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/util/ArrayList.html?is-external=true" title="class or interface in java.util">ArrayList</a>&lt;<a href="../../../../../../../com/compomics/util/experiment/identification/matches/IonMatch.html" title="class in com.compomics.util.experiment.identification.matches">IonMatch</a>&gt;&gt;&nbsp;allAnnotations,
                <a href="http://docs.oracle.com/javase/7/docs/api/java/util/ArrayList.html?is-external=true" title="class or interface in java.util">ArrayList</a>&lt;<a href="../../../../../../../com/compomics/util/experiment/mass_spectrometry/spectra/Spectrum.html" title="class in com.compomics.util.experiment.mass_spectrometry.spectra">Spectrum</a>&gt;&nbsp;allSpectra,
                <a href="http://docs.oracle.com/javase/7/docs/api/java/util/HashSet.html?is-external=true" title="class or interface in java.util">HashSet</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Integer.html?is-external=true" title="class or interface in java.lang">Integer</a>&gt;&nbsp;currentFragmentIonTypes,
                <a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/NeutralLossesMap.html" title="class in com.compomics.util.experiment.identification.spectrum_annotation">NeutralLossesMap</a>&nbsp;neutralLosses,
                boolean&nbsp;singleCharge,
                boolean&nbsp;twoCharges)</code>
<div class="block">Creates a novel fragment ion table displaying bar charts with the
 intensity of each fragment ion type.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colLast"><code><span class="memberNameLink"><a href="../../../../../../../com/compomics/util/gui/spectrum/FragmentIonTable.html#FragmentIonTable-com.compomics.util.experiment.biology.proteins.Peptide-java.util.ArrayList-java.util.HashSet-com.compomics.util.experiment.identification.spectrum_annotation.NeutralLossesMap-boolean-boolean-">FragmentIonTable</a></span>(<a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Peptide</a>&nbsp;currentPeptide,
                <a href="http://docs.oracle.com/javase/7/docs/api/java/util/ArrayList.html?is-external=true" title="class or interface in java.util">ArrayList</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/util/ArrayList.html?is-external=true" title="class or interface in java.util">ArrayList</a>&lt;<a href="../../../../../../../com/compomics/util/experiment/identification/matches/IonMatch.html" title="class in com.compomics.util.experiment.identification.matches">IonMatch</a>&gt;&gt;&nbsp;allAnnotations,
                <a href="http://docs.oracle.com/javase/7/docs/api/java/util/HashSet.html?is-external=true" title="class or interface in java.util">HashSet</a>&lt;<a href="http://docs.oracle.com/javase/7/docs/api/java/lang/Integer.html?is-external=true" title="class or interface in java.lang">Integer</a>&gt;&nbsp;currentFragmentIonTypes,
                <a href="../../../../../../../com/compomics/util/experiment/identification/spectrum_annotation/NeutralLossesMap.html" title="class in com.compomics.util.experiment.identification.spectrum_annotation">NeutralLossesMap</a>&nbsp;neutralLosses,
                boolean&nbsp;singleCharge,
                boolean&nbsp;twoCharges)</code>
<div class="block">Creates a traditional fragment ion table with the theoretical mz values
 and the detected fragment ions highlighted.</div>
</td>
</tr>
</tbody>
</table>
</li>
</ul>
</li>
</ul>
</div>
<!-- ======= START OF BOTTOM NAVBAR ====== -->
<div class="bottomNav"><a name="navbar.bottom">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.bottom" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.bottom.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../../../../../../../overview-summary.html">Overview</a></li>
<li><a href="../package-summary.html">Package</a></li>
<li><a href="../../../../../../../com/compomics/util/experiment/biology/proteins/Peptide.html" title="class in com.compomics.util.experiment.biology.proteins">Class</a></li>
<li class="navBarCell1Rev">Use</li>
<li><a href="../package-tree.html">Tree</a></li>
<li><a href="../../../../../../../deprecated-list.html">Deprecated</a></li>
<li><a href="../../../../../../../index-all.html">Index</a></li>
<li><a href="../../../../../../../help-doc.html">Help</a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li>Prev</li>
<li>Next</li>
</ul>
<ul class="navList">
<li><a href="../../../../../../../index.html?com/compomics/util/experiment/biology/proteins/class-use/Peptide.html" target="_top">Frames</a></li>
<li><a href="Peptide.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_bottom">
<li><a href="../../../../../../../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_bottom");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<a name="skip.navbar.bottom">
<!--   -->
</a></div>
<!-- ======== END OF BOTTOM NAVBAR ======= -->
<p class="legalCopy"><small>Copyright &#169; 2018. All rights reserved.</small></p>
</body>
</html>
